// @spartacus/user/esm2015/profile/components/close-account/close-account.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { IconModule, SpinnerModule } from '@spartacus/storefront';
import { CloseAccountModalComponent } from './components/close-account-modal/close-account-modal.component';
import { CloseAccountComponent } from './components/close-account/close-account.component';
export class CloseAccountModule {
}
CloseAccountModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    UrlModule,
                    I18nModule,
                    IconModule,
                    SpinnerModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            CloseAccountComponent: {
                                component: CloseAccountComponent,
                                guards: [AuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [CloseAccountComponent, CloseAccountModalComponent],
            },] }
];
