// @spartacus/storefront/esm2015/cms-components/product/product-intro/product-intro.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { StarRatingModule } from '../../../shared/components/star-rating/star-rating.module';
import { ProductIntroComponent } from './product-intro.component';
export class ProductIntroModule {
}
ProductIntroModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, I18nModule, StarRatingModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ProductIntroComponent: {
                                component: ProductIntroComponent,
                            },
                        },
                    }),
                ],
                declarations: [ProductIntroComponent],
                exports: [ProductIntroComponent],
                entryComponents: [ProductIntroComponent],
            },] }
];
