// @spartacus/storefront/esm2015/layout/theme/theme.service.js
import { Injectable, RendererFactory2, } from '@angular/core';
import { SiteContextConfig, THEME_CONTEXT_ID } from '@spartacus/core';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
export class ThemeService {
    constructor(config, rendererFactory) {
        this.config = config;
        this.rendererFactory = rendererFactory;
    }
    /**
     * This function is to be called for the root component that is
     * bootstrapped.
     */
    init(rootComponent) {
        var _a;
        this.renderer = this.rendererFactory.createRenderer(null, null);
        this.rootComponent = rootComponent;
        // Theme value is a string. It is put in the generic multi-value
        // property of the SiteContextConfig. So the array's first item
        // is the theme value.
        this.setTheme((_a = this.config.context[THEME_CONTEXT_ID]) === null || _a === void 0 ? void 0 : _a[0]);
    }
    setTheme(theme) {
        if (theme) {
            const element = this.rootComponent.location.nativeElement;
            // remove the old theme
            this.renderer.removeClass(element, this.existingTheme);
            // add the new theme
            this.renderer.addClass(element, theme);
            this.existingTheme = theme;
        }
    }
}
ThemeService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ThemeService_Factory() { return new ThemeService(i0.ɵɵinject(i1.SiteContextConfig), i0.ɵɵinject(i0.RendererFactory2)); }, token: ThemeService, providedIn: "root" });
ThemeService.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
ThemeService.ctorParameters = () => [
    { type: SiteContextConfig },
    { type: RendererFactory2 }
];
