// @spartacus/storefront/esm2015/cms-components/myaccount/order/replenishment-order-history/replenishment-order-history.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, I18nModule, provideConfig, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ListNavigationModule } from '../../../../shared/components/list-navigation/list-navigation.module';
import { CmsPageGuard } from '../../../../cms-structure/guards/cms-page.guard';
import { PageLayoutComponent } from '../../../../cms-structure/page/page-layout/page-layout.component';
import { ReplenishmentOrderHistoryComponent } from './replenishment-order-history.component';
import { defaultReplenishmentOrderCancellationLayoutConfig } from '../replenishment-order-details/default-replenishment-order-cancellation-layout.config';
const ɵ0 = { cxRoute: 'replenishmentOrders' };
export class ReplenishmentOrderHistoryModule {
}
ReplenishmentOrderHistoryModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    RouterModule.forChild([
                        {
                            path: null,
                            canActivate: [AuthGuard, CmsPageGuard],
                            component: PageLayoutComponent,
                            data: ɵ0,
                        },
                    ]),
                    RouterModule,
                    ListNavigationModule,
                    UrlModule,
                    I18nModule,
                ],
                providers: [
                    provideConfig(defaultReplenishmentOrderCancellationLayoutConfig),
                    provideDefaultConfig({
                        cmsComponents: {
                            AccountReplenishmentHistoryComponent: {
                                component: ReplenishmentOrderHistoryComponent,
                                guards: [AuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [ReplenishmentOrderHistoryComponent],
                exports: [ReplenishmentOrderHistoryComponent],
                entryComponents: [ReplenishmentOrderHistoryComponent],
            },] }
];
export { ɵ0 };
