// @spartacus/user/esm2015/profile/components/reset-password/reset-password-component.service.js
import { Injectable } from '@angular/core';
import { FormControl, FormGroup, Validators } from '@angular/forms';
import { GlobalMessageService, GlobalMessageType, RoutingService, } from '@spartacus/core';
import { CustomFormValidators } from '@spartacus/storefront';
import { UserPasswordFacade } from '@spartacus/user/profile/root';
import { BehaviorSubject } from 'rxjs';
import { map, tap } from 'rxjs/operators';
export class ResetPasswordComponentService {
    constructor(userPasswordService, routingService, globalMessage) {
        this.userPasswordService = userPasswordService;
        this.routingService = routingService;
        this.globalMessage = globalMessage;
        this.busy$ = new BehaviorSubject(false);
        this.isUpdating$ = this.busy$.pipe(tap((state) => (state === true ? this.form.disable() : this.form.enable())));
        this.resetToken$ = this.routingService
            .getRouterState()
            .pipe(map((routerState) => routerState.state.queryParams['token']));
        this.form = new FormGroup({
            password: new FormControl('', [
                Validators.required,
                CustomFormValidators.passwordValidator,
            ]),
            passwordConfirm: new FormControl('', Validators.required),
        }, {
            validators: CustomFormValidators.passwordsMustMatch('password', 'passwordConfirm'),
        });
    }
    /**
     * Resets the password by the given token.
     *
     * The token has been provided during the request password flow.
     * The token is not validated on the client.
     */
    resetPassword(token) {
        if (!this.form.valid) {
            this.form.markAllAsTouched();
            return;
        }
        this.busy$.next(true);
        const password = this.form.get('password').value;
        this.userPasswordService.reset(token, password).subscribe({
            next: () => this.onSuccess(),
            error: (error) => this.onError(error),
        });
    }
    onSuccess() {
        this.globalMessage.add({ key: 'forgottenPassword.passwordResetSuccess' }, GlobalMessageType.MSG_TYPE_CONFIRMATION);
        this.busy$.next(false);
        this.form.reset();
        this.redirect();
    }
    onError(error) {
        var _a, _b;
        this.busy$.next(false);
        ((_b = (_a = error) === null || _a === void 0 ? void 0 : _a.details) !== null && _b !== void 0 ? _b : []).forEach((err) => {
            if (err.message) {
                this.globalMessage.add({ raw: err.message }, GlobalMessageType.MSG_TYPE_ERROR);
            }
        });
    }
    /**
     * Redirects the user to the login page.
     */
    redirect() {
        this.routingService.go({ cxRoute: 'login' });
    }
}
ResetPasswordComponentService.decorators = [
    { type: Injectable }
];
ResetPasswordComponentService.ctorParameters = () => [
    { type: UserPasswordFacade },
    { type: RoutingService },
    { type: GlobalMessageService }
];
