// @spartacus/user/esm2015/profile/components/reset-password/reset-password.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { GlobalMessageService, I18nModule, NotAuthGuard, provideDefaultConfig, RoutingService, } from '@spartacus/core';
import { FormErrorsModule, SpinnerModule } from '@spartacus/storefront';
import { UserPasswordFacade } from '@spartacus/user/profile/root';
import { ResetPasswordComponentService } from './reset-password-component.service';
import { ResetPasswordComponent } from './reset-password.component';
export class ResetPasswordModule {
}
ResetPasswordModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    RouterModule,
                    I18nModule,
                    FormErrorsModule,
                    SpinnerModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ResetPasswordComponent: {
                                component: ResetPasswordComponent,
                                guards: [NotAuthGuard],
                                providers: [
                                    {
                                        provide: ResetPasswordComponentService,
                                        useClass: ResetPasswordComponentService,
                                        deps: [UserPasswordFacade, RoutingService, GlobalMessageService],
                                    },
                                ],
                            },
                        },
                    }),
                ],
                declarations: [ResetPasswordComponent],
            },] }
];
