// @spartacus/storefront/esm2015/events/cart/cart-page-event.builder.js
import { Injectable } from '@angular/core';
import { ActionsSubject } from '@ngrx/store';
import { createFrom, EventService, FeatureConfigService, } from '@spartacus/core';
import { filter, map } from 'rxjs/operators';
import { NavigationEvent } from '../navigation/navigation.event';
import { CartPageEvent } from './cart-page.events';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "@spartacus/core";
export class CartPageEventBuilder {
    constructor(
    // TODO: #10896 - remove this
    /** @deprecated @since 3.1 - this will be removed in 4.0 */
    actions, eventService, 
    // TODO: #10896 - remove this
    /** @deprecated @since 3.1 - this will be removed in 4.0 */ featureConfigService) {
        this.actions = actions;
        this.eventService = eventService;
        this.featureConfigService = featureConfigService;
        this.register();
    }
    register() {
        this.eventService.register(CartPageEvent, this.buildCartPageEvent());
    }
    buildCartPageEvent() {
        return this.eventService.get(NavigationEvent).pipe(filter((navigationEvent) => navigationEvent.semanticRoute === 'cart'), map((navigationEvent) => createFrom(CartPageEvent, Object.assign(Object.assign({}, this.createDeprecatedPageEvent(navigationEvent)), { navigation: Object.assign({}, navigationEvent) }))));
    }
    // TODO: #10896 - remove this method
    createDeprecatedPageEvent(navigationEvent) {
        if (!this.featureConfigService ||
            this.featureConfigService.isLevel('!3.1')) {
            return Object.assign({}, navigationEvent);
        }
        return undefined;
    }
}
CartPageEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartPageEventBuilder_Factory() { return new CartPageEventBuilder(i0.ɵɵinject(i1.ActionsSubject), i0.ɵɵinject(i2.EventService), i0.ɵɵinject(i2.FeatureConfigService)); }, token: CartPageEventBuilder, providedIn: "root" });
CartPageEventBuilder.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CartPageEventBuilder.ctorParameters = () => [
    { type: ActionsSubject },
    { type: EventService },
    { type: FeatureConfigService }
];
