// @spartacus/core/esm2015/src/cms/page/config/default-page-meta.config.js
export const defaultPageMetaConfig = {
    pageMeta: {
        resolvers: [
            {
                property: 'title',
                method: 'resolveTitle',
            },
            {
                property: 'heading',
                method: 'resolveHeading',
            },
            {
                property: 'breadcrumbs',
                method: 'resolveBreadcrumbs',
            },
            // @TODO (#10467) disable (`disabledInCsr`) description, image and robots resolvers to optimize resolving logic
            {
                property: 'description',
                method: 'resolveDescription',
            },
            {
                property: 'image',
                method: 'resolveImage',
            },
            {
                property: 'robots',
                method: 'resolveRobots',
            },
        ],
        canonicalUrl: {
            forceHttps: true,
            forceWww: false,
            removeQueryParams: true,
            forceTrailingSlash: true,
        },
    },
};
