// @spartacus/storefront/esm2015/cms-components/myaccount/close-account/components/close-account/close-account.component.js
import { ChangeDetectionStrategy, Component } from '@angular/core';
import { ModalService } from '../../../../../shared/components/modal/index';
import { CloseAccountModalComponent } from '../close-account-modal/close-account-modal.component';
/**
 * @deprecated since 3.2, moved to @spartacus/user package.
 */
export class CloseAccountComponent {
    constructor(modalService) {
        this.modalService = modalService;
    }
    openModal() {
        this.modal = this.modalService.open(CloseAccountModalComponent, {
            centered: true,
        }).componentInstance;
    }
}
CloseAccountComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-close-account',
                template: "<div class=\"col-lg-8 col-md-9\">\n  <div class=\"row cx-btn-group\">\n    <div class=\"col-sm-3\">\n      <a\n        [routerLink]=\"{ cxRoute: 'home' } | cxUrl\"\n        class=\"btn btn-block btn-secondary\"\n        >{{ 'common.cancel' | cxTranslate }}</a\n      >\n    </div>\n    <div class=\"col-sm-6\">\n      <button class=\"btn btn-primary\" (click)=\"openModal()\">\n        {{ 'closeAccount.closeMyAccount' | cxTranslate }}\n      </button>\n    </div>\n  </div>\n</div>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
CloseAccountComponent.ctorParameters = () => [
    { type: ModalService }
];
