// @spartacus/user/fesm2015/spartacus-user-profile-root.js
import { ɵɵdefineInjectable, Injectable, NgModule } from '@angular/core';
import { facadeFactory, provideDefaultConfigFactory, UserProfileFacadeTransitionalToken, UserRegisterFacadeTransitionalToken } from '@spartacus/core';

const USER_PROFILE_FEATURE = 'userProfile';
const USER_PROFILE_CORE_FEATURE = 'userProfileCore';

function UserProfileFacadeFactory() {
    return facadeFactory({
        facade: UserProfileFacade,
        feature: USER_PROFILE_CORE_FEATURE,
        methods: ['get', 'update', 'close', 'getTitles'],
    });
}
class UserProfileFacade {
}
UserProfileFacade.ɵprov = ɵɵdefineInjectable({ factory: UserProfileFacadeFactory, token: UserProfileFacade, providedIn: "root" });
UserProfileFacade.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
                useFactory: UserProfileFacadeFactory,
            },] }
];

function UserRegisterFacadeFactory() {
    return facadeFactory({
        facade: UserRegisterFacade,
        feature: USER_PROFILE_CORE_FEATURE,
        methods: ['register', 'registerGuest', 'getTitles'],
    });
}
class UserRegisterFacade {
}
UserRegisterFacade.ɵprov = ɵɵdefineInjectable({ factory: UserRegisterFacadeFactory, token: UserRegisterFacade, providedIn: "root" });
UserRegisterFacade.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
                useFactory: UserRegisterFacadeFactory,
            },] }
];

// TODO: Inline this factory when we start releasing Ivy compiled libraries
function defaultUserProfileComponentsConfig() {
    const config = {
        featureModules: {
            [USER_PROFILE_FEATURE]: {
                cmsComponents: [
                    'RegisterCustomerComponent',
                    'UpdateProfileComponent',
                    'UpdateEmailComponent',
                    'UpdatePasswordComponent',
                    'ForgotPasswordComponent',
                    'ResetPasswordComponent',
                    'CloseAccountComponent',
                ],
            },
            // by default core is bundled together with components
            [USER_PROFILE_CORE_FEATURE]: USER_PROFILE_FEATURE,
        },
    };
    return config;
}
class UserProfileRootModule {
}
UserProfileRootModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    provideDefaultConfigFactory(defaultUserProfileComponentsConfig),
                    {
                        provide: UserProfileFacadeTransitionalToken,
                        useExisting: UserProfileFacade,
                    },
                    {
                        provide: UserRegisterFacadeTransitionalToken,
                        useExisting: UserRegisterFacade,
                    },
                ],
            },] }
];

function UserEmailFacadeFactory() {
    return facadeFactory({
        facade: UserEmailFacade,
        feature: USER_PROFILE_CORE_FEATURE,
        methods: ['update'],
    });
}
class UserEmailFacade {
}
UserEmailFacade.ɵprov = ɵɵdefineInjectable({ factory: UserEmailFacadeFactory, token: UserEmailFacade, providedIn: "root" });
UserEmailFacade.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
                useFactory: UserEmailFacadeFactory,
            },] }
];

function UserPasswordFacadeFactory() {
    return facadeFactory({
        facade: UserPasswordFacade,
        feature: USER_PROFILE_CORE_FEATURE,
        methods: ['update', 'reset', 'requestForgotPasswordEmail'],
    });
}
class UserPasswordFacade {
}
UserPasswordFacade.ɵprov = ɵɵdefineInjectable({ factory: UserPasswordFacadeFactory, token: UserPasswordFacade, providedIn: "root" });
UserPasswordFacade.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
                useFactory: UserPasswordFacadeFactory,
            },] }
];

/** AUGMENTABLE_TYPES_END */

/**
 * Generated bundle index. Do not edit.
 */

export { USER_PROFILE_CORE_FEATURE, USER_PROFILE_FEATURE, UserEmailFacade, UserEmailFacadeFactory, UserPasswordFacade, UserPasswordFacadeFactory, UserProfileFacade, UserProfileFacadeFactory, UserProfileRootModule, UserRegisterFacade, UserRegisterFacadeFactory, defaultUserProfileComponentsConfig };
