// @spartacus/smartedit/esm2015/root/http-interceptors/cms-ticket.interceptor.js
import { Injectable } from '@angular/core';
import { SmartEditLauncherService } from '../services/smart-edit-launcher.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/smart-edit-launcher.service";
export class CmsTicketInterceptor {
    constructor(service) {
        this.service = service;
    }
    intercept(request, next) {
        if (this.service.cmsTicketId && request.url.includes('/cms/')) {
            request = request.clone({
                setParams: {
                    cmsTicketId: this.service.cmsTicketId,
                },
            });
        }
        return next.handle(request);
    }
}
CmsTicketInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function CmsTicketInterceptor_Factory() { return new CmsTicketInterceptor(i0.ɵɵinject(i1.SmartEditLauncherService)); }, token: CmsTicketInterceptor, providedIn: "root" });
CmsTicketInterceptor.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
CmsTicketInterceptor.ctorParameters = () => [
    { type: SmartEditLauncherService }
];
