// @spartacus/core/esm2015/src/user/connectors/replenishment-order/user-replenishment-order.connector.js
import { Injectable } from '@angular/core';
import { UserReplenishmentOrderAdapter } from './user-replenishment-order.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user-replenishment-order.adapter";
export class UserReplenishmentOrderConnector {
    constructor(adapter) {
        this.adapter = adapter;
    }
    load(userId, replenishmentOrderCode) {
        return this.adapter.load(userId, replenishmentOrderCode);
    }
    loadReplenishmentDetailsHistory(userId, replenishmentOrderCode, pageSize, currentPage, sort) {
        return this.adapter.loadReplenishmentDetailsHistory(userId, replenishmentOrderCode, pageSize, currentPage, sort);
    }
    cancelReplenishmentOrder(userId, replenishmentOrderCode) {
        return this.adapter.cancelReplenishmentOrder(userId, replenishmentOrderCode);
    }
    loadHistory(userId, pageSize, currentPage, sort) {
        return this.adapter.loadHistory(userId, pageSize, currentPage, sort);
    }
}
UserReplenishmentOrderConnector.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserReplenishmentOrderConnector_Factory() { return new UserReplenishmentOrderConnector(i0.ɵɵinject(i1.UserReplenishmentOrderAdapter)); }, token: UserReplenishmentOrderConnector, providedIn: "root" });
UserReplenishmentOrderConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
UserReplenishmentOrderConnector.ctorParameters = () => [
    { type: UserReplenishmentOrderAdapter }
];
