// @spartacus/core/esm2015/src/auth/client-auth/services/client-token.service.js
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { queueScheduler } from 'rxjs';
import { filter, map, observeOn } from 'rxjs/operators';
import { ClientAuthActions } from '../store/actions/index';
import { ClientAuthSelectors } from '../store/selectors/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
/**
 * Serves a role of a facade on client token store.
 */
export class ClientTokenService {
    constructor(store) {
        this.store = store;
    }
    /**
     * Returns a client token. The client token from the store is returned if there is one.
     * Otherwise a new token is fetched from the backend and saved in the store.
     */
    getClientToken() {
        return this.store.pipe(select(ClientAuthSelectors.getClientTokenState), observeOn(queueScheduler), filter((state) => {
            if (this.isClientTokenLoaded(state)) {
                return true;
            }
            else {
                if (!state.loading) {
                    this.store.dispatch(new ClientAuthActions.LoadClientToken());
                }
                return false;
            }
        }), map((state) => state.value));
    }
    /**
     * Fetches a clientToken from the backend and saves it in the store where getClientToken can use it.
     * The new clientToken is returned.
     */
    refreshClientToken() {
        this.store.dispatch(new ClientAuthActions.LoadClientToken());
        return this.store.pipe(select(ClientAuthSelectors.getClientTokenState), filter((state) => this.isClientTokenLoaded(state)), map((state) => state.value));
    }
    isClientTokenLoaded(state) {
        return (state.success || state.error) && !state.loading;
    }
}
ClientTokenService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ClientTokenService_Factory() { return new ClientTokenService(i0.ɵɵinject(i1.Store)); }, token: ClientTokenService, providedIn: "root" });
ClientTokenService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ClientTokenService.ctorParameters = () => [
    { type: Store }
];
