// @spartacus/core/esm2015/src/auth/client-auth/services/client-error-handling.service.js
import { Injectable } from '@angular/core';
import { switchMap, take } from 'rxjs/operators';
import { ClientTokenService } from './client-token.service';
import * as i0 from "@angular/core";
import * as i1 from "./client-token.service";
/**
 * Service for handling `Authorization` header and errors for requests that
 * require client token (eg. user registration).
 */
export class ClientErrorHandlingService {
    constructor(clientTokenService) {
        this.clientTokenService = clientTokenService;
    }
    /**
     * Refreshes client token and retries the request with the new token.
     *
     * @param request
     * @param httpHandler
     */
    handleExpiredClientToken(request, next) {
        return this.clientTokenService.refreshClientToken().pipe(take(1), switchMap((token) => {
            return next.handle(this.createNewRequestWithNewToken(request, token));
        }));
    }
    /**
     * Clones the requests and provided `Authorization` header.
     *
     * @param request
     * @param token
     */
    createNewRequestWithNewToken(request, token) {
        request = request.clone({
            setHeaders: {
                Authorization: `${token.token_type || 'Bearer'} ${token.access_token}`,
            },
        });
        return request;
    }
}
ClientErrorHandlingService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ClientErrorHandlingService_Factory() { return new ClientErrorHandlingService(i0.ɵɵinject(i1.ClientTokenService)); }, token: ClientErrorHandlingService, providedIn: "root" });
ClientErrorHandlingService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ClientErrorHandlingService.ctorParameters = () => [
    { type: ClientTokenService }
];
