// @spartacus/storefront/esm2015/cms-components/content/paragraph/paragraph.module.js
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { ParagraphComponent } from './paragraph.component';
import { provideDefaultConfig } from '@spartacus/core';
export class CmsParagraphModule {
}
CmsParagraphModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            CMSParagraphComponent: {
                                component: ParagraphComponent,
                            },
                            CMSTabParagraphComponent: {
                                component: ParagraphComponent,
                            },
                        },
                    }),
                ],
                declarations: [ParagraphComponent],
                exports: [ParagraphComponent],
                entryComponents: [ParagraphComponent],
            },] }
];
