// @spartacus/storefront/esm2015/shared/components/popover/popover.service.js
import { Injectable } from '@angular/core';
import * as i0 from "@angular/core";
export class PopoverService {
    /**
     * For a11y improvements method returns different `FocusConfig`
     * based on which event popover was triggered.
     */
    getFocusConfig(event, appendToBody) {
        let config = {};
        if (event instanceof KeyboardEvent) {
            if ((event.code === 'Enter' || event.code === 'Space') && appendToBody) {
                config = {
                    trap: true,
                    block: true,
                    focusOnEscape: false,
                    autofocus: true,
                };
            }
        }
        return config;
    }
    setFocusOnElement(element, focusConfig, appendToBody) {
        if (focusConfig && appendToBody) {
            element.nativeElement.focus();
        }
    }
}
PopoverService.ɵprov = i0.ɵɵdefineInjectable({ factory: function PopoverService_Factory() { return new PopoverService(); }, token: PopoverService, providedIn: "root" });
PopoverService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
