// @spartacus/core/esm2015/src/checkout/store/actions/replenishment-order.action.js
import { PROCESS_FEATURE } from '../../../process/store/process-state';
import { StateUtils } from '../../../state/utils/index';
import { PLACED_ORDER_PROCESS_ID } from '../checkout-state';
export const SCHEDULE_REPLENISHMENT_ORDER = '[Checkout] Schedule Replenishment Order';
export const SCHEDULE_REPLENISHMENT_ORDER_SUCCESS = '[Checkout] Schedule Replenishment Order Success';
export const SCHEDULE_REPLENISHMENT_ORDER_FAIL = '[Checkout] Schedule Replenishment Order Fail';
export const CLEAR_SCHEDULE_REPLENISHMENT_ORDER = '[Checkout] Clear Schedule Replenishment Data';
export class ScheduleReplenishmentOrder extends StateUtils.EntityLoadAction {
    constructor(payload) {
        super(PROCESS_FEATURE, PLACED_ORDER_PROCESS_ID);
        this.payload = payload;
        this.type = SCHEDULE_REPLENISHMENT_ORDER;
    }
}
export class ScheduleReplenishmentOrderSuccess extends StateUtils.EntitySuccessAction {
    constructor(payload) {
        super(PROCESS_FEATURE, PLACED_ORDER_PROCESS_ID);
        this.payload = payload;
        this.type = SCHEDULE_REPLENISHMENT_ORDER_SUCCESS;
    }
}
export class ScheduleReplenishmentOrderFail extends StateUtils.EntityFailAction {
    constructor(payload) {
        super(PROCESS_FEATURE, PLACED_ORDER_PROCESS_ID, payload);
        this.payload = payload;
        this.type = SCHEDULE_REPLENISHMENT_ORDER_FAIL;
    }
}
export class ClearScheduleReplenishmentOrderAction extends StateUtils.EntityLoaderResetAction {
    constructor() {
        super(PROCESS_FEATURE, PLACED_ORDER_PROCESS_ID);
        this.type = CLEAR_SCHEDULE_REPLENISHMENT_ORDER;
    }
}
