// @spartacus/storefront/esm2015/cms-components/myaccount/order/amend-order/returns/order-return.guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { SemanticPathService } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { OrderReturnService } from './order-return.service';
import * as i0 from "@angular/core";
import * as i1 from "./order-return.service";
import * as i2 from "@spartacus/core";
import * as i3 from "@angular/router";
export class OrderReturnGuard {
    constructor(orderAmendService, semanticPathService, router) {
        this.orderAmendService = orderAmendService;
        this.semanticPathService = semanticPathService;
        this.router = router;
    }
    canActivate() {
        return this.orderAmendService.getForm().pipe(map((form) => {
            if (!form.valid) {
                // the order code is not available in the route
                // as long as we're inside a guard, hence we redirect
                // to the common orders page.
                return this.router.parseUrl(this.semanticPathService.get('orders'));
            }
            else {
                return true;
            }
        }));
    }
}
OrderReturnGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function OrderReturnGuard_Factory() { return new OrderReturnGuard(i0.ɵɵinject(i1.OrderReturnService), i0.ɵɵinject(i2.SemanticPathService), i0.ɵɵinject(i3.Router)); }, token: OrderReturnGuard, providedIn: "root" });
OrderReturnGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
OrderReturnGuard.ctorParameters = () => [
    { type: OrderReturnService },
    { type: SemanticPathService },
    { type: Router }
];
