// @spartacus/storefront/esm2015/shared/components/form/form-errors/form-errors.component.js
import { ChangeDetectionStrategy, Component, HostBinding, Input, } from '@angular/core';
import { map, startWith } from 'rxjs/operators';
/**
 * This component renders form errors.
 */
export class FormErrorsComponent {
    constructor() {
        this.prefix = 'formErrors';
    }
    set control(control) {
        this._control = control;
        this.errors$ = control === null || control === void 0 ? void 0 : control.statusChanges.pipe(startWith({}), map(() => control.errors || {}), map((errors) => Object.entries(errors)
            .filter((error) => error[1])
            .map((error) => error[0])));
    }
    get control() {
        return this._control;
    }
    get invalid() {
        var _a;
        return (_a = this.control) === null || _a === void 0 ? void 0 : _a.invalid;
    }
    get dirty() {
        var _a;
        return (_a = this.control) === null || _a === void 0 ? void 0 : _a.dirty;
    }
    get touched() {
        var _a;
        return (_a = this.control) === null || _a === void 0 ? void 0 : _a.touched;
    }
}
FormErrorsComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-form-errors',
                template: "<p *ngFor=\"let errorName of errors$ | async\">\n  {{ prefix + '.' + errorName | cxTranslate: translationParams }}\n</p>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
FormErrorsComponent.propDecorators = {
    prefix: [{ type: Input }],
    translationParams: [{ type: Input }],
    control: [{ type: Input }],
    invalid: [{ type: HostBinding, args: ['class.control-invalid',] }],
    dirty: [{ type: HostBinding, args: ['class.control-dirty',] }],
    touched: [{ type: HostBinding, args: ['class.control-touched',] }]
};
