// @spartacus/storefront/esm2015/cms-components/myaccount/payment-methods/payment-methods.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { CardModule } from '../../../shared/components/card/card.module';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { PaymentMethodsComponent } from './payment-methods.component';
export class PaymentMethodsModule {
}
PaymentMethodsModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, CardModule, SpinnerModule, I18nModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            AccountPaymentDetailsComponent: {
                                component: PaymentMethodsComponent,
                                guards: [AuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [PaymentMethodsComponent],
                exports: [PaymentMethodsComponent],
                entryComponents: [PaymentMethodsComponent],
            },] }
];
