// @spartacus/smartedit/esm2015/root/services/smart-edit-launcher.service.js
import { Location } from '@angular/common';
import { Injectable } from '@angular/core';
import { FeatureModulesService } from '@spartacus/core';
import { SmartEditConfig } from '../config/smart-edit-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/smart-edit-config";
import * as i2 from "@angular/common";
import * as i3 from "@spartacus/core";
/**
 * The SmartEditLauncherService is used to check whether Spartacus is launched inside Smart Edit;
 * it also gets cmsTicketId sent from Smart Edit.
 */
export class SmartEditLauncherService {
    constructor(config, location, featureModules) {
        this.config = config;
        this.location = location;
        this.featureModules = featureModules;
    }
    get cmsTicketId() {
        return this._cmsTicketId;
    }
    /**
     * Lazy loads modules when Spartacus launced inside Smart Edit
     */
    load() {
        if (this.isLaunchedInSmartEdit() &&
            this.featureModules.isConfigured('smartEdit')) {
            this.featureModules.resolveFeature('smartEdit').subscribe();
        }
    }
    /**
     * Indicates whether Spartacus is launched in SmartEdit
     */
    isLaunchedInSmartEdit() {
        const path = this.location.path().split('?')[0];
        const params = this.location.path().split('?')[1];
        const cmsToken = params === null || params === void 0 ? void 0 : params.split('&').find((param) => param.startsWith('cmsTicketId='));
        this._cmsTicketId = cmsToken === null || cmsToken === void 0 ? void 0 : cmsToken.split('=')[1];
        return (path.split('/').pop() === this.config.smartEdit.storefrontPreviewRoute &&
            !!this._cmsTicketId);
    }
}
SmartEditLauncherService.ɵprov = i0.ɵɵdefineInjectable({ factory: function SmartEditLauncherService_Factory() { return new SmartEditLauncherService(i0.ɵɵinject(i1.SmartEditConfig), i0.ɵɵinject(i2.Location), i0.ɵɵinject(i3.FeatureModulesService)); }, token: SmartEditLauncherService, providedIn: "root" });
SmartEditLauncherService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
SmartEditLauncherService.ctorParameters = () => [
    { type: SmartEditConfig },
    { type: Location },
    { type: FeatureModulesService }
];
