// @spartacus/storefront/esm2015/cms-components/misc/qualtrics/qualtrics.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { defaultQualtricsConfig } from './config/default-qualtrics-config';
import { QualtricsComponent } from './qualtrics.component';
/**
 * @deprecated since 3.1 - moved to feature-lib
 * Please take a look at https://sap.github.io/spartacus-docs/qualtrics-integration/#page-title
 * to see how to migrate into the new feature-lib.
 * Do not import from the storefront. Instead import from the qualtrics feature-lib.
 */
export class QualtricsModule {
}
QualtricsModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                declarations: [QualtricsComponent],
                entryComponents: [QualtricsComponent],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            QualtricsComponent: {
                                component: QualtricsComponent,
                            },
                        },
                    }),
                    provideDefaultConfig(defaultQualtricsConfig),
                ],
            },] }
];
