// @spartacus/user/esm2015/account/core/facade/user-account.service.js
import { Injectable } from '@angular/core';
import { UserAccountChangedEvent, } from '@spartacus/user/account/root';
import { LoginEvent, LogoutEvent, QueryService, UserIdService, } from '@spartacus/core';
import { switchMap } from 'rxjs/operators';
import { UserAccountConnector } from '../connectors/user-account.connector';
export class UserAccountService {
    constructor(userAccountConnector, userIdService, query) {
        this.userAccountConnector = userAccountConnector;
        this.userIdService = userIdService;
        this.query = query;
        this.userQuery = this.query.create(() => this.userIdService
            .takeUserId(true)
            .pipe(switchMap((userId) => this.userAccountConnector.get(userId))), {
            reloadOn: [UserAccountChangedEvent],
            resetOn: [LoginEvent, LogoutEvent],
        });
    }
    /**
     * Returns the current user.
     */
    get() {
        return this.userQuery.get();
    }
}
UserAccountService.decorators = [
    { type: Injectable }
];
UserAccountService.ctorParameters = () => [
    { type: UserAccountConnector },
    { type: UserIdService },
    { type: QueryService }
];
