// @spartacus/storefront/esm2015/cms-components/myaccount/order/order-details/order-detail-items/order-consigned-entries/order-consigned-entries.component.js
import { Component, Input } from '@angular/core';
import { PromotionLocation, } from '@spartacus/core';
export class OrderConsignedEntriesComponent {
    constructor() {
        this.promotionLocation = PromotionLocation.Order;
    }
    getConsignmentProducts(consignment) {
        const products = [];
        consignment.entries.forEach((element) => {
            products.push(element.orderEntry);
        });
        return products;
    }
}
OrderConsignedEntriesComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-order-consigned-entries',
                template: "<div *ngFor=\"let consignment of consignments\" class=\"cx-list row\">\n  <div class=\"cx-list-header col-12\">\n    <div class=\"cx-list-status\">\n      <span *ngIf=\"consignment\">\n        {{\n          'orderDetails.deliveryStatus'\n            | cxTranslate: { context: consignment.status }\n        }}\n      </span>\n    </div>\n    <div *ngIf=\"consignment?.statusDate\" class=\"cx-list-date\">\n      <div>{{ consignment?.statusDate | cxDate }}</div>\n    </div>\n\n    <cx-consignment-tracking\n      [orderCode]=\"order.code\"\n      [consignment]=\"consignment\"\n      *cxFeature=\"'consignmentTracking'\"\n    >\n    </cx-consignment-tracking>\n  </div>\n  <div class=\"cx-list-item col-12\">\n    <cx-cart-item-list\n      [items]=\"consignment.entries\"\n      [readonly]=\"true\"\n      [promotionLocation]=\"promotionLocation\"\n    ></cx-cart-item-list>\n  </div>\n</div>\n"
            },] }
];
OrderConsignedEntriesComponent.propDecorators = {
    consignments: [{ type: Input }],
    order: [{ type: Input }]
};
