// @spartacus/storefront/esm2015/layout/launch-dialog/launch-dialog.module.js
import { NgModule } from '@angular/core';
import { Config } from '@spartacus/core';
import { LayoutConfig } from '../config/layout-config';
import { InlineRenderStrategy, LaunchRenderStrategy, OutletRenderStrategy, RoutingRenderStrategy, } from './services/index';
export class LaunchDialogModule {
    static forRoot() {
        return {
            ngModule: LaunchDialogModule,
            providers: [{ provide: LayoutConfig, useExisting: Config }],
        };
    }
}
LaunchDialogModule.decorators = [
    { type: NgModule, args: [{
                providers: [
                    {
                        provide: LaunchRenderStrategy,
                        useExisting: OutletRenderStrategy,
                        multi: true,
                    },
                    {
                        provide: LaunchRenderStrategy,
                        useExisting: InlineRenderStrategy,
                        multi: true,
                    },
                    {
                        provide: LaunchRenderStrategy,
                        useExisting: RoutingRenderStrategy,
                        multi: true,
                    },
                ],
            },] }
];
