// @spartacus/core/esm2015/src/checkout/services/clear-checkout.service.js
import { Injectable } from '@angular/core';
import { CheckoutDeliveryService } from '../facade/checkout-delivery.service';
import { CheckoutPaymentService } from '../facade/checkout-payment.service';
import * as i0 from "@angular/core";
import * as i1 from "../facade/checkout-delivery.service";
import * as i2 from "../facade/checkout-payment.service";
export class ClearCheckoutService {
    constructor(checkoutDeliveryService, checkoutPaymentService) {
        this.checkoutDeliveryService = checkoutDeliveryService;
        this.checkoutPaymentService = checkoutPaymentService;
    }
    /**
     * Use it to clear checkout state
     */
    resetCheckoutProcesses() {
        this.checkoutDeliveryService.resetSetDeliveryAddressProcess();
        this.checkoutDeliveryService.resetSetDeliveryModeProcess();
        this.checkoutPaymentService.resetSetPaymentDetailsProcess();
    }
}
ClearCheckoutService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ClearCheckoutService_Factory() { return new ClearCheckoutService(i0.ɵɵinject(i1.CheckoutDeliveryService), i0.ɵɵinject(i2.CheckoutPaymentService)); }, token: ClearCheckoutService, providedIn: "root" });
ClearCheckoutService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ClearCheckoutService.ctorParameters = () => [
    { type: CheckoutDeliveryService },
    { type: CheckoutPaymentService }
];
