// @spartacus/storefront/esm2015/cms-components/checkout/components/payment-type/payment-type.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ConfigModule, I18nModule } from '@spartacus/core';
import { SpinnerModule } from '../../../../shared/components/spinner/spinner.module';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { CartNotEmptyGuard } from './../../../../cms-components/cart/cart-not-empty.guard';
import { PaymentTypeComponent } from './payment-type.component';
export class PaymentTypeModule {
}
PaymentTypeModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    I18nModule,
                    SpinnerModule,
                    ConfigModule.withConfig({
                        cmsComponents: {
                            CheckoutPaymentType: {
                                component: PaymentTypeComponent,
                                guards: [CheckoutAuthGuard, CartNotEmptyGuard],
                            },
                        },
                    }),
                ],
                declarations: [PaymentTypeComponent],
                entryComponents: [PaymentTypeComponent],
                exports: [PaymentTypeComponent],
            },] }
];
