// @spartacus/core/esm2015/src/occ/config-loader/occ-sites-config-loader.js
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { map } from 'rxjs/operators';
import { OccConfig } from '../config/occ-config';
import * as i0 from "@angular/core";
import * as i1 from "../config/occ-config";
import * as i2 from "@angular/common/http";
/**
 * @deprecated since 3.2, use `SiteConnector` instead
 */
// TODO(#11515): drop it in 4.0
export class OccSitesConfigLoader {
    constructor(config, http) {
        this.config = config;
        this.http = http;
        this.endpoint = 'basesites?fields=baseSites(uid,defaultLanguage(isocode),urlEncodingAttributes,urlPatterns,stores(currencies(isocode),defaultCurrency(isocode),languages(isocode),defaultLanguage(isocode)),theme)';
    }
    getPrefix() {
        var _a, _b, _c, _d, _e;
        if (Boolean((_b = (_a = this.config.backend) === null || _a === void 0 ? void 0 : _a.occ) === null || _b === void 0 ? void 0 : _b.prefix) &&
            !((_e = (_d = (_c = this.config.backend) === null || _c === void 0 ? void 0 : _c.occ) === null || _d === void 0 ? void 0 : _d.prefix) === null || _e === void 0 ? void 0 : _e.startsWith('/'))) {
            return '/' + this.config.backend.occ.prefix;
        }
        return this.config.backend.occ.prefix;
    }
    get baseEndpoint() {
        return (this.config.backend.occ.baseUrl || '') + this.getPrefix();
    }
    get url() {
        return `${this.baseEndpoint}${this.endpoint}`;
    }
    load() {
        if (!this.config || !this.config.backend || !this.config.backend.occ) {
            return throwError(new Error(`Missing config for OCC backend!`));
        }
        return this.http
            .get(this.url)
            .pipe(map(({ baseSites }) => baseSites));
    }
}
OccSitesConfigLoader.ɵprov = i0.ɵɵdefineInjectable({ factory: function OccSitesConfigLoader_Factory() { return new OccSitesConfigLoader(i0.ɵɵinject(i1.OccConfig), i0.ɵɵinject(i2.HttpClient)); }, token: OccSitesConfigLoader, providedIn: "root" });
OccSitesConfigLoader.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
OccSitesConfigLoader.ctorParameters = () => [
    { type: OccConfig },
    { type: HttpClient }
];
