// @spartacus/user/esm2015/profile/core/facade/user-profile.service.js
import { Injectable } from '@angular/core';
import { AuthService, CommandService, CommandStrategy, EventService, LanguageSetEvent, QueryService, UserIdService, } from '@spartacus/core';
import { UserAccountChangedEvent, UserAccountFacade, } from '@spartacus/user/account/root';
import { UserProfileConnector } from '../connectors/user-profile.connector';
import { map, switchMap, tap } from 'rxjs/operators';
export class UserProfileService {
    constructor(userAccountService, authService, userProfileConnector, eventService, userIdService, query, command) {
        this.userAccountService = userAccountService;
        this.authService = authService;
        this.userProfileConnector = userProfileConnector;
        this.eventService = eventService;
        this.userIdService = userIdService;
        this.query = query;
        this.command = command;
        this.updateCommand = this.command.create((payload) => this.userIdService.takeUserId(true).pipe(switchMap((uid) => this.userProfileConnector.update(uid, payload.details).pipe(tap(() => {
            this.eventService.dispatch({ user: payload.details }, UserAccountChangedEvent);
        })))), {
            strategy: CommandStrategy.Queue,
        });
        this.closeCommand = this.command.create(() => this.userIdService
            .takeUserId(true)
            .pipe(switchMap((uid) => this.userProfileConnector
            .remove(uid)
            .pipe(tap(() => this.authService.logout())))));
        this.titleQuery = this.query.create(() => this.userProfileConnector.getTitles(), {
            reloadOn: [LanguageSetEvent],
        });
    }
    get() {
        return this.userAccountService.get();
    }
    /**
     * Updates the user's details.
     *
     * @param details User details to be updated.
     */
    update(details) {
        return this.updateCommand.execute({ details });
    }
    /**
     * Closes the user account.
     */
    close() {
        return this.closeCommand.execute(undefined);
    }
    /**
     * Returns titles that can be used for the user profiles.
     */
    getTitles() {
        return this.titleQuery.get().pipe(map((titles) => titles !== null && titles !== void 0 ? titles : []));
    }
}
UserProfileService.decorators = [
    { type: Injectable }
];
UserProfileService.ctorParameters = () => [
    { type: UserAccountFacade },
    { type: AuthService },
    { type: UserProfileConnector },
    { type: EventService },
    { type: UserIdService },
    { type: QueryService },
    { type: CommandService }
];
