// @spartacus/user/esm2015/profile/core/facade/user-email.service.js
import { Injectable } from '@angular/core';
import { CommandService, CommandStrategy, UserIdService, } from '@spartacus/core';
import { switchMap } from 'rxjs/operators';
import { UserProfileConnector } from '../connectors/user-profile.connector';
export class UserEmailService {
    constructor(userIdService, userProfileConnector, command) {
        this.userIdService = userIdService;
        this.userProfileConnector = userProfileConnector;
        this.command = command;
        this.updateCommand = this.command.create((payload) => this.userIdService
            .takeUserId(true)
            .pipe(switchMap((uid) => this.userProfileConnector.updateEmail(uid, payload.password, payload.newUid))), {
            strategy: CommandStrategy.Queue,
        });
    }
    /**
     * Updates the user's email.
     *
     * @param password to users password to confirm the users
     * @param newUid the new proposed email address.
     */
    update(password, newUid) {
        return this.updateCommand.execute({ password, newUid });
    }
}
UserEmailService.decorators = [
    { type: Injectable }
];
UserEmailService.ctorParameters = () => [
    { type: UserIdService },
    { type: UserProfileConnector },
    { type: CommandService }
];
