// @spartacus/storefront/esm2015/cms-components/myaccount/reset-password/reset-password.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, } from '@spartacus/core';
import { FormErrorsModule } from '../../../shared/index';
import { ResetPasswordFormComponent } from './reset-password-form.component';
/**
 * @deprecated since 3.2, moved to @spartacus/user package.
 */
export class ResetPasswordModule {
}
ResetPasswordModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    RouterModule,
                    I18nModule,
                    FormErrorsModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ResetPasswordComponent: {
                                component: ResetPasswordFormComponent,
                                guards: [NotAuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [ResetPasswordFormComponent],
                exports: [ResetPasswordFormComponent],
                entryComponents: [ResetPasswordFormComponent],
            },] }
];
