// @spartacus/user/esm2015/profile/components/update-password/update-password.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { AuthGuard, GlobalMessageService, I18nModule, provideDefaultConfig, RoutingService, UrlModule, } from '@spartacus/core';
import { FormErrorsModule, SpinnerModule } from '@spartacus/storefront';
import { UserPasswordFacade } from '@spartacus/user/profile/root';
import { UpdatePasswordComponentService } from './update-password-component.service';
import { UpdatePasswordComponent } from './update-password.component';
export class UpdatePasswordModule {
}
UpdatePasswordModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    SpinnerModule,
                    I18nModule,
                    FormErrorsModule,
                    UrlModule,
                    RouterModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            UpdatePasswordComponent: {
                                component: UpdatePasswordComponent,
                                guards: [AuthGuard],
                                providers: [
                                    {
                                        provide: UpdatePasswordComponentService,
                                        useClass: UpdatePasswordComponentService,
                                        deps: [UserPasswordFacade, RoutingService, GlobalMessageService],
                                    },
                                ],
                            },
                        },
                    }),
                ],
                declarations: [UpdatePasswordComponent],
            },] }
];
