// @spartacus/core/esm2015/src/cms/page/content-page-meta.resolver.js
import { Injectable } from '@angular/core';
import { combineLatest, defer, of } from 'rxjs';
import { filter, map, shareReplay } from 'rxjs/operators';
import { TranslationService } from '../../i18n/translation.service';
import { PageType } from '../../model/cms.model';
import { CmsService } from '../facade/cms.service';
import { BasePageMetaResolver } from './base-page-meta.resolver';
import { PageMetaResolver } from './page-meta.resolver';
import { RoutingPageMetaResolver } from './routing/routing-page-meta.resolver';
import * as i0 from "@angular/core";
import * as i1 from "../facade/cms.service";
import * as i2 from "../../i18n/translation.service";
import * as i3 from "./routing/routing-page-meta.resolver";
import * as i4 from "./base-page-meta.resolver";
/**
 * Resolves the page data for all Content Pages based on the `PageType.CONTENT_PAGE`.
 * More specific resolvers for content pages can be implemented by making them more
 * specific, for example by using the page template (see `CartPageMetaResolver`).
 *
 * The page title, and breadcrumbs are resolved in this implementation only.
 */
export class ContentPageMetaResolver extends PageMetaResolver {
    constructor(cmsService, translation, routingPageMetaResolver, basePageMetaResolver) {
        super();
        this.cmsService = cmsService;
        this.translation = translation;
        this.routingPageMetaResolver = routingPageMetaResolver;
        this.basePageMetaResolver = basePageMetaResolver;
        /**
         * Breadcrumb for the home page.
         *
         * @deprecated since 3.1, as we resolve the homeBreadcrumb$ from the `BasePageMetaResolver`
         */
        // TODO(#10467): drop the homeBreadcrumb$ property
        this.homeBreadcrumb$ = this.translation
            .translate('common.home')
            .pipe(map((label) => [{ label: label, link: '/' }]));
        /**
         * All the resolved breadcrumbs (including those from Angular child routes).
         *
         * @deprecated since 3.1, as we resolve the breadcrumbs$ from the `BasePageMetaResolver`
         */
        // TODO(#10467): drop the breadcrumbs$ property
        this.breadcrumbs$ = combineLatest([
            this.homeBreadcrumb$,
            defer(() => this.routingPageMetaResolver.resolveBreadcrumbs()),
        ]).pipe(map((breadcrumbs) => breadcrumbs.flat()), shareReplay({ bufferSize: 1, refCount: true }));
        /**
         * Helper to provide access to the current CMS page
         *
         * @deprecated since 3.1, as we resolve the cms page data from the `BasePageMetaResolver`
         */
        // TODO(#10467): drop the cms$ property
        this.cms$ = defer(() => this.cmsService.getCurrentPage().pipe(filter((p) => Boolean(p))));
        /**
         * @deprecated since 3.1, we'll start using the `BasePageMetaResolver` to resolve
         * the page title
         */
        // TODO(#10467): drop the title$ property
        this.title$ = this.cms$.pipe(map((p) => p.title));
        this.pageType = PageType.CONTENT_PAGE;
    }
    // TODO(#10467): resolve the title from the `BasePageMetaResolver.resolveTitle()` only
    resolveTitle() {
        return this.basePageMetaResolver
            ? this.basePageMetaResolver.resolveTitle()
            : this.title$;
    }
    resolveDescription() {
        return this.basePageMetaResolver
            ? this.basePageMetaResolver.resolveDescription()
            : of();
    }
    /**
     * @override
     * Resolves a single breadcrumb item to the home page for each `ContentPage`.
     * The home page label is resolved from the translation service.
     */
    resolveBreadcrumbs() {
        return this.basePageMetaResolver
            ? this.basePageMetaResolver.resolveBreadcrumbs()
            : this.breadcrumbs$;
    }
    /**
     * @override
     * This is added in 3.1 and will be ignored if the `BasePageMetaResolver` is not
     * available.
     */
    // TODO(#10467) drop the 3.1 note.
    resolveRobots() {
        var _a, _b;
        return (_b = (_a = this.basePageMetaResolver) === null || _a === void 0 ? void 0 : _a.resolveRobots()) !== null && _b !== void 0 ? _b : of();
    }
    /**
     * @override resolves the canonical page for the content page.
     */
    resolveCanonicalUrl() {
        var _a, _b;
        return (_b = (_a = this.basePageMetaResolver) === null || _a === void 0 ? void 0 : _a.resolveCanonicalUrl()) !== null && _b !== void 0 ? _b : of();
    }
}
ContentPageMetaResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function ContentPageMetaResolver_Factory() { return new ContentPageMetaResolver(i0.ɵɵinject(i1.CmsService), i0.ɵɵinject(i2.TranslationService), i0.ɵɵinject(i3.RoutingPageMetaResolver), i0.ɵɵinject(i4.BasePageMetaResolver)); }, token: ContentPageMetaResolver, providedIn: "root" });
ContentPageMetaResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ContentPageMetaResolver.ctorParameters = () => [
    { type: CmsService },
    { type: TranslationService },
    { type: RoutingPageMetaResolver },
    { type: BasePageMetaResolver }
];
