// @spartacus/storefront/esm2015/cms-components/product/product-summary/product-summary.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { OutletModule } from '../../../cms-structure/outlet/outlet.module';
import { ProductSummaryComponent } from './product-summary.component';
export class ProductSummaryModule {
}
ProductSummaryModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, OutletModule, I18nModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ProductSummaryComponent: {
                                component: ProductSummaryComponent,
                            },
                        },
                    }),
                ],
                declarations: [ProductSummaryComponent],
                entryComponents: [ProductSummaryComponent],
                exports: [ProductSummaryComponent],
            },] }
];
