// @spartacus/storefront/esm2015/cms-components/checkout/guards/checkout-auth.guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { ActiveCartService, AuthRedirectService, AuthService, B2BUserRole, GlobalMessageService, GlobalMessageType, SemanticPathService, UserService, } from '@spartacus/core';
import { combineLatest } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { CheckoutConfigService } from '../services/checkout-config.service';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "../services/checkout-config.service";
import * as i3 from "@angular/router";
export class CheckoutAuthGuard {
    constructor(authService, authRedirectService, checkoutConfigService, activeCartService, semanticPathService, router, userService, globalMessageService) {
        this.authService = authService;
        this.authRedirectService = authRedirectService;
        this.checkoutConfigService = checkoutConfigService;
        this.activeCartService = activeCartService;
        this.semanticPathService = semanticPathService;
        this.router = router;
        this.userService = userService;
        this.globalMessageService = globalMessageService;
    }
    canActivate() {
        return combineLatest([
            this.authService.isUserLoggedIn(),
            this.activeCartService.getAssignedUser(),
            this.userService.get(),
            this.activeCartService.isStable(),
        ]).pipe(filter(([, , , isStable]) => Boolean(isStable)), map(([isLoggedIn, cartUser, user]) => {
            if (!isLoggedIn) {
                if (this.activeCartService.isGuestCart()) {
                    return Boolean(cartUser);
                }
                this.authRedirectService.reportAuthGuard();
                if (this.checkoutConfigService.isGuestCheckout()) {
                    return this.router.createUrlTree([this.semanticPathService.get('login')], { queryParams: { forced: true } });
                }
                else {
                    return this.router.parseUrl(this.semanticPathService.get('login'));
                }
            }
            else if ('roles' in user) {
                const roles = user.roles;
                if (roles.includes(B2BUserRole.CUSTOMER)) {
                    return true;
                }
                this.globalMessageService.add({ key: 'checkout.invalid.accountType' }, GlobalMessageType.MSG_TYPE_WARNING);
                return false;
            }
            return isLoggedIn;
        }));
    }
}
CheckoutAuthGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function CheckoutAuthGuard_Factory() { return new CheckoutAuthGuard(i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.AuthRedirectService), i0.ɵɵinject(i2.CheckoutConfigService), i0.ɵɵinject(i1.ActiveCartService), i0.ɵɵinject(i1.SemanticPathService), i0.ɵɵinject(i3.Router), i0.ɵɵinject(i1.UserService), i0.ɵɵinject(i1.GlobalMessageService)); }, token: CheckoutAuthGuard, providedIn: "root" });
CheckoutAuthGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CheckoutAuthGuard.ctorParameters = () => [
    { type: AuthService },
    { type: AuthRedirectService },
    { type: CheckoutConfigService },
    { type: ActiveCartService },
    { type: SemanticPathService },
    { type: Router },
    { type: UserService },
    { type: GlobalMessageService }
];
