// @spartacus/storefront/esm2015/cms-components/checkout/guards/not-checkout-auth.guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { ActiveCartService, AuthService, SemanticPathService, } from '@spartacus/core';
import { map } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "@angular/router";
export class NotCheckoutAuthGuard {
    constructor(authService, activeCartService, semanticPathService, router) {
        this.authService = authService;
        this.activeCartService = activeCartService;
        this.semanticPathService = semanticPathService;
        this.router = router;
    }
    canActivate() {
        return this.authService.isUserLoggedIn().pipe(map((isLoggedIn) => {
            if (isLoggedIn) {
                return this.router.parseUrl(this.semanticPathService.get('home'));
            }
            else if (this.activeCartService.isGuestCart()) {
                return this.router.parseUrl(this.semanticPathService.get('cart'));
            }
            return !isLoggedIn;
        }));
    }
}
NotCheckoutAuthGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function NotCheckoutAuthGuard_Factory() { return new NotCheckoutAuthGuard(i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.ActiveCartService), i0.ɵɵinject(i1.SemanticPathService), i0.ɵɵinject(i2.Router)); }, token: NotCheckoutAuthGuard, providedIn: "root" });
NotCheckoutAuthGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
NotCheckoutAuthGuard.ctorParameters = () => [
    { type: AuthService },
    { type: ActiveCartService },
    { type: SemanticPathService },
    { type: Router }
];
