// @spartacus/storefront/esm2015/cms-components/myaccount/order/replenishment-order-details/replenishment-order-cancellation/replenishment-order-cancellation.component.js
import { Component, ViewChild, ViewContainerRef, } from '@angular/core';
import { UserReplenishmentOrderService, } from '@spartacus/core';
import { LaunchDialogService, } from '../../../../../layout/launch-dialog';
import { Subscription } from 'rxjs';
import { take } from 'rxjs/operators';
import { ReplenishmentOrderCancellationLaunchDialogService } from './replenishment-order-cancellation-launch-dialog.service';
export class ReplenishmentOrderCancellationComponent {
    constructor(userReplenishmentOrderService, replenishmentOrderCancellationLaunchDialogService, vcr, launchDialogService) {
        this.userReplenishmentOrderService = userReplenishmentOrderService;
        this.replenishmentOrderCancellationLaunchDialogService = replenishmentOrderCancellationLaunchDialogService;
        this.vcr = vcr;
        this.launchDialogService = launchDialogService;
        this.subscription = new Subscription();
        this.replenishmentOrder$ = this.userReplenishmentOrderService.getReplenishmentOrderDetails();
    }
    openDialog() {
        // TODO(#12167): use launchDialogService only
        if (this.launchDialogService) {
            const dialog = this.launchDialogService.openDialog("REPLENISHMENT_ORDER" /* REPLENISHMENT_ORDER */, this.element, this.vcr);
            if (dialog) {
                this.subscription.add(dialog.pipe(take(1)).subscribe());
            }
        }
        else {
            const dialog = this.replenishmentOrderCancellationLaunchDialogService.openDialog(this.element, this.vcr);
            if (dialog) {
                this.subscription.add(dialog.pipe(take(1)).subscribe());
            }
        }
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
        this.userReplenishmentOrderService.clearReplenishmentOrderDetails();
    }
}
ReplenishmentOrderCancellationComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-replenishment-order-cancellation',
                template: "<div class=\"cx-cancel-replenishment-btns row\">\n  <div class=\"col-xs-12 col-md-5 col-lg-4\">\n    <a\n      class=\"btn btn-block btn-action\"\n      [routerLink]=\"\n        {\n          cxRoute: 'replenishmentOrders'\n        } | cxUrl\n      \"\n    >\n      {{ 'common.back' | cxTranslate }}\n    </a>\n  </div>\n  <div\n    *ngIf=\"(replenishmentOrder$ | async).active\"\n    class=\"col-xs-12 col-md-5 col-lg-4\"\n  >\n    <button #element class=\"btn btn-block btn-action\" (click)=\"openDialog()\">\n      {{ 'orderDetails.cancelReplenishment.title' | cxTranslate }}\n    </button>\n  </div>\n</div>\n"
            },] }
];
ReplenishmentOrderCancellationComponent.ctorParameters = () => [
    { type: UserReplenishmentOrderService },
    { type: ReplenishmentOrderCancellationLaunchDialogService },
    { type: ViewContainerRef },
    { type: LaunchDialogService }
];
ReplenishmentOrderCancellationComponent.propDecorators = {
    element: [{ type: ViewChild, args: ['element',] }]
};
