// @spartacus/core/esm2015/src/auth/user-auth/events/user-auth-event.builder.js
import { Injectable } from '@angular/core';
import { filter, map, pairwise } from 'rxjs/operators';
import { EventService } from '../../../event/event.service';
import { StateEventService } from '../../../state/event/index';
import { createFrom } from '../../../util/create-from';
import { AuthService } from '../facade/auth.service';
import { AuthActions } from '../store/actions/index';
import { LoginEvent, LogoutEvent } from './user-auth.events';
import * as i0 from "@angular/core";
import * as i1 from "../../../state/event/state-event.service";
import * as i2 from "../../../event/event.service";
import * as i3 from "../facade/auth.service";
export class UserAuthEventBuilder {
    constructor(stateEventService, eventService, authService) {
        this.stateEventService = stateEventService;
        this.eventService = eventService;
        this.authService = authService;
        this.register();
    }
    /**
     * Registers user auth events
     */
    register() {
        this.registerLoginEvent();
        this.registerLogoutEvent();
    }
    /**
     * Register a login event
     */
    registerLoginEvent() {
        this.stateEventService.register({
            action: AuthActions.LOGIN,
            event: LoginEvent,
        });
    }
    /**
     * Register a logout event
     */
    registerLogoutEvent() {
        this.eventService.register(LogoutEvent, this.buildLogoutEvent());
    }
    /**
     * Returns logout event stream
     */
    buildLogoutEvent() {
        return this.authService.isUserLoggedIn().pipe(pairwise(), filter(([prev, curr]) => prev && !curr), map(() => createFrom(LogoutEvent, {})));
    }
}
UserAuthEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserAuthEventBuilder_Factory() { return new UserAuthEventBuilder(i0.ɵɵinject(i1.StateEventService), i0.ɵɵinject(i2.EventService), i0.ɵɵinject(i3.AuthService)); }, token: UserAuthEventBuilder, providedIn: "root" });
UserAuthEventBuilder.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
UserAuthEventBuilder.ctorParameters = () => [
    { type: StateEventService },
    { type: EventService },
    { type: AuthService }
];
