// @spartacus/core/esm2015/src/user/connectors/user/user.connector.js
import { Injectable } from '@angular/core';
import { UserAdapter } from './user.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./user.adapter";
export class UserConnector {
    constructor(adapter) {
        this.adapter = adapter;
    }
    get(userId) {
        return this.adapter.load(userId);
    }
    update(username, user) {
        return this.adapter.update(username, user);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    register(user) {
        return this.adapter.register(user);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    registerGuest(guid, password) {
        return this.adapter.registerGuest(guid, password);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    requestForgotPasswordEmail(userEmailAddress) {
        return this.adapter.requestForgotPasswordEmail(userEmailAddress);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    resetPassword(token, newPassword) {
        return this.adapter.resetPassword(token, newPassword);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    updateEmail(userId, currentPassword, newUserId) {
        return this.adapter.updateEmail(userId, currentPassword, newUserId);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    updatePassword(userId, oldPassword, newPassword) {
        return this.adapter.updatePassword(userId, oldPassword, newPassword);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    remove(userId) {
        return this.adapter.remove(userId);
    }
    /**
     * @deprecated since 3.2, use `UserProfileConnector` from `@spartacus/user` package
     */
    getTitles() {
        return this.adapter.loadTitles();
    }
}
UserConnector.ɵprov = i0.ɵɵdefineInjectable({ factory: function UserConnector_Factory() { return new UserConnector(i0.ɵɵinject(i1.UserAdapter)); }, token: UserConnector, providedIn: "root" });
UserConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
UserConnector.ctorParameters = () => [
    { type: UserAdapter }
];
