// @spartacus/storefront/esm2015/recipes/b2c-storefront.module.js
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { provideConfig, provideDefaultConfig } from '@spartacus/core';
import { CmsLibModule } from '../cms-components/cms-lib.module';
import { layoutConfig, mediaConfig } from './config/index';
import { defaultCmsContentProviders } from './config/static-cms-structure';
import { StorefrontModule } from './storefront.module';
export class B2cStorefrontModule {
    static withConfig(config) {
        return {
            ngModule: B2cStorefrontModule,
            providers: [provideConfig(config)],
        };
    }
}
B2cStorefrontModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                    StorefrontModule,
                    // the cms lib module contains all components that added in the bundle
                    CmsLibModule,
                ],
                providers: [
                    provideDefaultConfig({
                        pwa: {
                            enabled: true,
                            addToHomeScreen: true,
                        },
                    }),
                    provideDefaultConfig(layoutConfig),
                    provideDefaultConfig(mediaConfig),
                    ...defaultCmsContentProviders,
                ],
                exports: [StorefrontModule],
            },] }
];
