// @spartacus/storefront/esm2015/cms-components/user/login-register/login-register.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { CheckoutModule, I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { PageSlotModule } from '../../../cms-structure/page/slot/page-slot.module';
import { LoginRegisterComponent } from './login-register.component';
/**
 * @deprecated since 3.2, use @spartacus/user package instead.
 */
export class LoginRegisterModule {
}
LoginRegisterModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    UrlModule,
                    PageSlotModule,
                    I18nModule,
                    CheckoutModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ReturningCustomerRegisterComponent: {
                                component: LoginRegisterComponent,
                                guards: [NotAuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [LoginRegisterComponent],
                entryComponents: [LoginRegisterComponent],
                exports: [LoginRegisterComponent],
            },] }
];
