// @spartacus/user/esm2015/account/components/login-register/login-register.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { PageSlotModule } from '@spartacus/storefront';
import { LoginRegisterComponent } from './login-register.component';
export class LoginRegisterModule {
}
LoginRegisterModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, RouterModule, UrlModule, PageSlotModule, I18nModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            ReturningCustomerRegisterComponent: {
                                component: LoginRegisterComponent,
                                guards: [NotAuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [LoginRegisterComponent],
            },] }
];
