// @spartacus/storefront/esm2015/cms-components/content/banner-carousel/banner-carousel.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { PageComponentModule } from '../../../cms-structure/page/component/page-component.module';
import { CarouselModule, MediaModule } from '../../../shared/components/index';
import { BannerCarouselComponent } from './banner-carousel.component';
export class BannerCarouselModule {
}
BannerCarouselModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, PageComponentModule, CarouselModule, MediaModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            RotatingImagesComponent: {
                                component: BannerCarouselComponent,
                            },
                        },
                    }),
                ],
                declarations: [BannerCarouselComponent],
                entryComponents: [BannerCarouselComponent],
                exports: [BannerCarouselComponent],
            },] }
];
