// @spartacus/core/esm2015/src/checkout/events/checkout-event.builder.js
import { Injectable } from '@angular/core';
import { StateEventService } from '../../state/event/state-event.service';
import { CheckoutActions } from '../store/actions/index';
import { OrderPlacedEvent } from './checkout.events';
import * as i0 from "@angular/core";
import * as i1 from "../../state/event/state-event.service";
export class CheckoutEventBuilder {
    constructor(stateEventService) {
        this.stateEventService = stateEventService;
        this.register();
    }
    /**
     * Registers checkout events
     */
    register() {
        this.orderPlacedEvent();
    }
    /**
     * Register an order successfully placed event
     */
    orderPlacedEvent() {
        this.stateEventService.register({
            action: CheckoutActions.PLACE_ORDER_SUCCESS,
            event: OrderPlacedEvent,
        });
    }
}
CheckoutEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function CheckoutEventBuilder_Factory() { return new CheckoutEventBuilder(i0.ɵɵinject(i1.StateEventService)); }, token: CheckoutEventBuilder, providedIn: "root" });
CheckoutEventBuilder.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CheckoutEventBuilder.ctorParameters = () => [
    { type: StateEventService }
];
