// @spartacus/storefront/esm2015/shared/components/modal/modal.directive.js
import { Directive, HostListener, Injectable, Input, Optional, } from '@angular/core';
import { Router, RouterLink, RouterLinkWithHref } from '@angular/router';
import { NgbActiveModal } from '@ng-bootstrap/ng-bootstrap';
import * as i0 from "@angular/core";
/**
 * Service to control the behavior of the DismissModalDirective
 */
export class ModalDirectiveService {
    /**
     * Handler for the click event on the directive
     *
     * @param options.type    Action to perform after click - close or dismiss
     * @param options.reason  Reason for dismissing or closing the modal
     * @param url             url (optional) - exists only when the directive is also a `routerLink`.
     */
    onClick(options, activeModal, url) {
        let reason = options.reason;
        if (!reason && url) {
            reason = `Link click: ${url}`;
        }
        if (options.type === 'dismiss') {
            activeModal.dismiss(reason);
        }
        else if (options.type === 'close') {
            activeModal.close(reason);
        }
    }
}
ModalDirectiveService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ModalDirectiveService_Factory() { return new ModalDirectiveService(); }, token: ModalDirectiveService, providedIn: "root" });
ModalDirectiveService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
/**
 * Directive to dismiss the modal on click
 */
export class ModalDirective {
    constructor(service, router, 
    // active modal can be injected only when the directive is projected inside modal
    activeModal, routerLink, routerLinkWithHref) {
        this.service = service;
        this.router = router;
        this.activeModal = activeModal;
        this.routerLink = routerLink;
        this.routerLinkWithHref = routerLinkWithHref;
    }
    /**
     * Returns URL in case when the directive is in the scope of a `routerLink` directive at the same time.
     */
    getUrl() {
        var _a;
        const routerLink = (_a = this.routerLink) !== null && _a !== void 0 ? _a : this.routerLinkWithHref;
        return routerLink
            ? this.router.serializeUrl(routerLink.urlTree)
            : undefined;
    }
    onClick() {
        if (this.activeModal) {
            this.service.onClick(this.options, this.activeModal, this.getUrl());
        }
    }
    /**
     * Converts the directive's inputs to the shape of modal directive options
     */
    get options() {
        return { type: this.cxModal, reason: this.cxModalReason };
    }
}
ModalDirective.decorators = [
    { type: Directive, args: [{
                selector: '[cxModal]',
            },] }
];
ModalDirective.ctorParameters = () => [
    { type: ModalDirectiveService },
    { type: Router },
    { type: NgbActiveModal, decorators: [{ type: Optional }] },
    { type: RouterLink, decorators: [{ type: Optional }] },
    { type: RouterLinkWithHref, decorators: [{ type: Optional }] }
];
ModalDirective.propDecorators = {
    cxModal: [{ type: Input }],
    cxModalReason: [{ type: Input }],
    onClick: [{ type: HostListener, args: ['click',] }]
};
