// @spartacus/storefront/esm2015/shared/components/modal/modal.service.js
import { ApplicationRef, Injectable } from '@angular/core';
import { NgbModal } from '@ng-bootstrap/ng-bootstrap';
import { FeatureConfigService } from '@spartacus/core';
import * as i0 from "@angular/core";
import * as i1 from "@ng-bootstrap/ng-bootstrap";
import * as i2 from "@spartacus/core";
/**
 * A service to handle modal
 */
export class ModalService {
    constructor(ngbModalService, 
    // TODO: make this param required in 4.0
    applicationRef, 
    // TODO: drop this param in 4.0
    featureConfigService) {
        this.ngbModalService = ngbModalService;
        this.applicationRef = applicationRef;
        this.featureConfigService = featureConfigService;
        this.modals = [];
    }
    get rootComponent() {
        var _a, _b, _c, _d;
        return (_d = (_c = (_b = (_a = this.applicationRef) === null || _a === void 0 ? void 0 : _a.components) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.location) === null || _d === void 0 ? void 0 : _d.nativeElement;
    }
    open(content, options) {
        var _a;
        let activeModal;
        // TODO: make this logic default in 4.0
        if ((_a = this.featureConfigService) === null || _a === void 0 ? void 0 : _a.isLevel('3.3')) {
            options = Object.assign({ container: this.rootComponent }, options);
        }
        activeModal = this.ngbModalService.open(content, options);
        this.modals.push(activeModal);
        this.handleModalRemoveEvents(activeModal);
        return activeModal;
    }
    handleModalRemoveEvents(modal) {
        modal.result.finally(() => {
            this.modals = this.modals.filter((m) => m !== modal);
        });
    }
    getActiveModal() {
        const modal = this.modals[this.modals.length - 1];
        return modal ? modal : null;
    }
    dismissActiveModal(reason) {
        const modal = this.getActiveModal();
        if (modal) {
            modal.dismiss(reason);
        }
    }
    closeActiveModal(reason) {
        const modal = this.getActiveModal();
        if (modal) {
            modal.close(reason);
        }
    }
}
ModalService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ModalService_Factory() { return new ModalService(i0.ɵɵinject(i1.NgbModal), i0.ɵɵinject(i0.ApplicationRef), i0.ɵɵinject(i2.FeatureConfigService)); }, token: ModalService, providedIn: "root" });
ModalService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ModalService.ctorParameters = () => [
    { type: NgbModal },
    { type: ApplicationRef },
    { type: FeatureConfigService }
];
