// @spartacus/core/esm2015/src/cart/store/selectors/multi-cart.selector.js
import { createFeatureSelector, createSelector, } from '@ngrx/store';
import { entityValueSelector } from '../../../state/utils/entity-loader/entity-loader.selectors';
import { entityHasPendingProcessesSelector, entityIsStableSelector, entityProcessesLoaderStateSelector, } from '../../../state/utils/entity-processes-loader/entity-processes-loader.selectors';
import { MULTI_CART_FEATURE, } from '../multi-cart-state';
export const getMultiCartState = createFeatureSelector(MULTI_CART_FEATURE);
const ɵ0 = (state) => state.carts;
export const getMultiCartEntities = createSelector(getMultiCartState, ɵ0);
export const getCartEntitySelectorFactory = (cartId) => {
    return createSelector(getMultiCartEntities, (state) => entityProcessesLoaderStateSelector(state, cartId));
};
export const getCartSelectorFactory = (cartId) => {
    return createSelector(getMultiCartEntities, (state) => entityValueSelector(state, cartId));
};
export const getCartIsStableSelectorFactory = (cartId) => {
    return createSelector(getMultiCartEntities, (state) => entityIsStableSelector(state, cartId));
};
export const getCartHasPendingProcessesSelectorFactory = (cartId) => {
    return createSelector(getMultiCartEntities, (state) => entityHasPendingProcessesSelector(state, cartId));
};
export const getCartEntriesSelectorFactory = (cartId) => {
    return createSelector(getCartSelectorFactory(cartId), (state) => {
        return state && state.entries ? state.entries : [];
    });
};
export const getCartEntrySelectorFactory = (cartId, productCode) => {
    return createSelector(getCartEntriesSelectorFactory(cartId), (state) => {
        return state
            ? state.find((entry) => entry.product.code === productCode)
            : undefined;
    });
};
const ɵ1 = (state) => state.active;
export const getActiveCartId = createSelector(getMultiCartState, ɵ1);
const ɵ2 = (state) => state.wishList;
export const getWishListId = createSelector(getMultiCartState, ɵ2);
const ɵ3 = (state) => Object.keys(state.entities).map((key) => entityValueSelector(state, key));
export const getCartsSelectorFactory = createSelector(getMultiCartEntities, ɵ3);
export { ɵ0, ɵ1, ɵ2, ɵ3 };
