// @spartacus/core/esm2015/src/cart/connectors/entry/cart-entry.connector.js
import { Injectable } from '@angular/core';
import { CartEntryAdapter } from './cart-entry.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./cart-entry.adapter";
export class CartEntryConnector {
    constructor(adapter) {
        this.adapter = adapter;
    }
    add(userId, cartId, productCode, quantity) {
        return this.adapter.add(userId, cartId, productCode, quantity);
    }
    update(userId, cartId, entryNumber, qty, pickupStore) {
        return this.adapter.update(userId, cartId, entryNumber, qty, pickupStore);
    }
    remove(userId, cartId, entryNumber) {
        return this.adapter.remove(userId, cartId, entryNumber);
    }
}
CartEntryConnector.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartEntryConnector_Factory() { return new CartEntryConnector(i0.ɵɵinject(i1.CartEntryAdapter)); }, token: CartEntryConnector, providedIn: "root" });
CartEntryConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CartEntryConnector.ctorParameters = () => [
    { type: CartEntryAdapter }
];
