// @spartacus/core/esm2015/src/occ/adapters/cart/occ-save-cart.adapter.js
import { HttpClient, HttpHeaders, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { SAVE_CART_NORMALIZER } from '../../../cart/connectors/save-cart/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
/**
 * TODO(#11938): 4.0: use OccSavedCartAdapter instead from saved-cart feature-lib of cart
 *
 * When the saved cart feature module exist, there exist a race condition between the 'saveCart' key for the default occ config
 * Allows wishlist to function when the key value is pulled from the saved-cart feature-lib
 *
 * Keep in mind that this adapter will be removed in 4.0 as we will delegate this functionality to the saved-cart feature-lib
 **/
export class OccSaveCartAdapter {
    constructor(http, occEndpointsService, converterService) {
        this.http = http;
        this.occEndpointsService = occEndpointsService;
        this.converterService = converterService;
    }
    saveCart(userId, cartId, saveCartName, saveCartDescription) {
        let httpParams = new HttpParams();
        if (Boolean(saveCartName)) {
            httpParams = httpParams.set('saveCartName', saveCartName);
        }
        if (Boolean(saveCartDescription)) {
            httpParams = httpParams.set('saveCartDescription', saveCartDescription);
        }
        const headers = new HttpHeaders({
            'Content-Type': 'application/x-www-form-urlencoded',
        });
        return !this.occEndpointsService
            .getUrl('saveCart', { userId, cartId })
            .includes('saveCartName')
            ? this.http
                .patch(this.occEndpointsService.getUrl('saveCart', { userId, cartId }), httpParams, { headers })
                .pipe(this.converterService.pipeable(SAVE_CART_NORMALIZER))
            : this.http
                .patch(this.occEndpointsService.getUrl('saveCart', {
                userId,
                cartId,
                saveCartName,
                saveCartDescription,
            }), cartId)
                .pipe(this.converterService.pipeable(SAVE_CART_NORMALIZER));
    }
}
OccSaveCartAdapter.decorators = [
    { type: Injectable }
];
OccSaveCartAdapter.ctorParameters = () => [
    { type: HttpClient },
    { type: OccEndpointsService },
    { type: ConverterService }
];
