// @spartacus/core/esm2015/src/cart/services/cart-page-meta.resolver.js
import { Injectable, Optional } from '@angular/core';
import { of } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { CmsService } from '../../cms/facade/cms.service';
import { PageRobotsMeta } from '../../cms/model/page.model';
import { BasePageMetaResolver } from '../../cms/page/base-page-meta.resolver';
import { PageMetaResolver } from '../../cms/page/page-meta.resolver';
import { PageType } from '../../model/cms.model';
import * as i0 from "@angular/core";
import * as i1 from "../../cms/facade/cms.service";
import * as i2 from "../../cms/page/base-page-meta.resolver";
/**
 * Resolves the page metadata for the Cart page (Using the `PageType.CONTENT_PAGE`
 * and the `CartPageTemplate`). If the cart page matches this template, the more
 * generic `ContentPageMetaResolver` is overridden by this resolver.
 *
 * The page title and robots are resolved in this implementation only.
 *
 * @deprecated since 3.1, in future versions we'll drop this service as the logic
 * is no longer specific since we introduce backend driven robots.
 */
// TODO(#10467): Remove implementation
export class CartPageMetaResolver extends PageMetaResolver {
    constructor(cms, basePageMetaResolver) {
        super();
        this.cms = cms;
        this.basePageMetaResolver = basePageMetaResolver;
        // TODO(#10467): remove the cms property as it's no longer needed when we use the `BasePageMetaResolver`
        /**
         * @deprecated since 3.1, we'll use the BasePageMetaResolver to resolve the page title
         */
        this.cms$ = this.cms
            .getCurrentPage()
            .pipe(filter((page) => !!page));
        this.pageType = PageType.CONTENT_PAGE;
        this.pageTemplate = 'CartPageTemplate';
    }
    resolveTitle() {
        // TODO(#10467): resolve the title from the `BasePageMetaResolver.resolveTitle()` only
        return this.basePageMetaResolver
            ? this.basePageMetaResolver.resolveTitle()
            : this.cms$.pipe(map((p) => p.title));
    }
    /**
     * @Override Returns robots for the cart pages, which default to NOINDEX/NOFOLLOW.
     */
    // TODO(#10467): resolve robots from `BasePageMetaResolver` instead
    resolveRobots() {
        return of([PageRobotsMeta.NOFOLLOW, PageRobotsMeta.NOINDEX]);
    }
}
CartPageMetaResolver.ɵprov = i0.ɵɵdefineInjectable({ factory: function CartPageMetaResolver_Factory() { return new CartPageMetaResolver(i0.ɵɵinject(i1.CmsService), i0.ɵɵinject(i2.BasePageMetaResolver)); }, token: CartPageMetaResolver, providedIn: "root" });
CartPageMetaResolver.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CartPageMetaResolver.ctorParameters = () => [
    { type: CmsService },
    { type: BasePageMetaResolver, decorators: [{ type: Optional }] }
];
