// @spartacus/storefront/esm2015/cms-components/checkout/components/cost-center/cost-center.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ConfigModule, I18nModule } from '@spartacus/core';
import { CartNotEmptyGuard } from '../../../cart';
import { CheckoutAuthGuard } from '../../guards';
import { CostCenterComponent } from './cost-center.component';
export class CostCenterModule {
}
CostCenterModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    I18nModule,
                    ConfigModule.withConfig({
                        cmsComponents: {
                            CheckoutCostCenterComponent: {
                                component: CostCenterComponent,
                                guards: [CheckoutAuthGuard, CartNotEmptyGuard],
                            },
                        },
                    }),
                ],
                declarations: [CostCenterComponent],
                entryComponents: [CostCenterComponent],
            },] }
];
