// @spartacus/storefront/esm2015/cms-components/cart/mini-cart/mini-cart.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { IconModule } from '../../../cms-components/misc/icon/icon.module';
import { MiniCartComponent } from './mini-cart.component';
export class MiniCartModule {
}
MiniCartModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, RouterModule, UrlModule, IconModule, I18nModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            MiniCartComponent: {
                                component: MiniCartComponent,
                            },
                        },
                    }),
                ],
                declarations: [MiniCartComponent],
                exports: [MiniCartComponent],
                entryComponents: [MiniCartComponent],
            },] }
];
