// @spartacus/storefront/esm2015/cms-components/checkout/components/checkout-order-summary/checkout-order-summary.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { provideDefaultConfig } from '@spartacus/core';
import { CartSharedModule } from '../../../../cms-components/cart/cart-shared/cart-shared.module';
import { CheckoutOrderSummaryComponent } from './checkout-order-summary.component';
export class CheckoutOrderSummaryModule {
}
CheckoutOrderSummaryModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, CartSharedModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            CheckoutOrderSummary: {
                                component: CheckoutOrderSummaryComponent,
                            },
                        },
                    }),
                ],
                declarations: [CheckoutOrderSummaryComponent],
                entryComponents: [CheckoutOrderSummaryComponent],
                exports: [CheckoutOrderSummaryComponent],
            },] }
];
