// @spartacus/storefront/esm2015/cms-components/myaccount/notification-preference/notification-preference.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { AuthGuard, I18nModule, provideDefaultConfig, } from '@spartacus/core';
import { SpinnerModule } from '../../../shared/components/spinner/spinner.module';
import { NotificationPreferenceComponent } from './notification-preference.component';
export class NotificationPreferenceModule {
}
NotificationPreferenceModule.decorators = [
    { type: NgModule, args: [{
                declarations: [NotificationPreferenceComponent],
                imports: [CommonModule, SpinnerModule, I18nModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            NotificationPreferenceComponent: {
                                component: NotificationPreferenceComponent,
                                guards: [AuthGuard],
                            },
                        },
                    }),
                ],
                exports: [NotificationPreferenceComponent],
                entryComponents: [NotificationPreferenceComponent],
            },] }
];
