// @spartacus/storefront/esm2015/cms-pages/product-listing-page/product-listing-page.module.js
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { provideDefaultConfig } from '@spartacus/core';
import { CmsPageGuard, PageLayoutComponent } from '../../cms-structure/index';
import { PRODUCT_LISTING_URL_MATCHER } from './product-listing-url-matcher';
const ɵ0 = { pageLabel: 'search', cxRoute: 'search' }, ɵ1 = { cxRoute: 'brand' }, ɵ2 = { cxRoute: 'category' };
export class ProductListingPageModule {
}
ProductListingPageModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    RouterModule.forChild([
                        {
                            path: null,
                            canActivate: [CmsPageGuard],
                            component: PageLayoutComponent,
                            data: ɵ0,
                        },
                        {
                            path: null,
                            canActivate: [CmsPageGuard],
                            component: PageLayoutComponent,
                            data: ɵ1,
                        },
                        {
                            // The 'category' route  may include a greedy suffix url matcher '**/c/:categoryCode'
                            // So not to shadow the specific 'brand' route, the 'category' is the last route in the sequence.
                            path: null,
                            canActivate: [CmsPageGuard],
                            component: PageLayoutComponent,
                            data: ɵ2,
                        },
                    ]),
                ],
                providers: [
                    provideDefaultConfig({
                        routing: {
                            routes: {
                                category: {
                                    matchers: [PRODUCT_LISTING_URL_MATCHER],
                                },
                            },
                        },
                    }),
                ],
            },] }
];
export { ɵ0, ɵ1, ɵ2 };
