// @spartacus/core/esm2015/src/auth/user-auth/http-interceptors/token-revocation.interceptor.js
import { Injectable } from '@angular/core';
import { switchMap, take } from 'rxjs/operators';
import { AuthConfigService } from '../services/auth-config.service';
import { AuthStorageService } from '../services/auth-storage.service';
import * as i0 from "@angular/core";
import * as i1 from "../services/auth-storage.service";
import * as i2 from "../services/auth-config.service";
/**
 * This interceptor is dedicated for Hybris OAuth server which requires `Authorization` header for revoke token calls.
 */
export class TokenRevocationInterceptor {
    constructor(authStorageService, authConfigService) {
        this.authStorageService = authStorageService;
        this.authConfigService = authConfigService;
    }
    intercept(request, next) {
        const isTokenRevocationRequest = this.isTokenRevocationRequest(request);
        return this.authStorageService.getToken().pipe(take(1), switchMap((token) => {
            if (isTokenRevocationRequest) {
                request = request.clone({
                    setHeaders: {
                        Authorization: `${token.token_type || 'Bearer'} ${token.access_token}`,
                    },
                });
            }
            return next.handle(request);
        }));
    }
    isTokenRevocationRequest(request) {
        return request.url === this.authConfigService.getRevokeEndpoint();
    }
}
TokenRevocationInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function TokenRevocationInterceptor_Factory() { return new TokenRevocationInterceptor(i0.ɵɵinject(i1.AuthStorageService), i0.ɵɵinject(i2.AuthConfigService)); }, token: TokenRevocationInterceptor, providedIn: "root" });
TokenRevocationInterceptor.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
TokenRevocationInterceptor.ctorParameters = () => [
    { type: AuthStorageService },
    { type: AuthConfigService }
];
