// @spartacus/storefront/esm2015/cms-components/checkout/components/schedule-replenishment-order/schedule-replenishment-order.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig } from '@spartacus/core';
import { CartNotEmptyGuard } from '../../../cart/cart-not-empty.guard';
import { IconModule } from '../../../misc/icon/icon.module';
import { CheckoutAuthGuard } from '../../guards/checkout-auth.guard';
import { ScheduleReplenishmentOrderComponent } from './schedule-replenishment-order.component';
export class ScheduleReplenishmentOrderModule {
}
ScheduleReplenishmentOrderModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, RouterModule, I18nModule, IconModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            CheckoutScheduleReplenishmentOrder: {
                                component: ScheduleReplenishmentOrderComponent,
                                guards: [CheckoutAuthGuard, CartNotEmptyGuard],
                            },
                        },
                    }),
                ],
                declarations: [ScheduleReplenishmentOrderComponent],
                entryComponents: [ScheduleReplenishmentOrderComponent],
                exports: [ScheduleReplenishmentOrderComponent],
            },] }
];
