// @spartacus/core/esm2015/src/asm/services/asm-auth-http-header.service.js
import { Injectable } from '@angular/core';
import { take } from 'rxjs/operators';
import { AuthService } from '../../auth/user-auth/facade/auth.service';
import { AuthHttpHeaderService } from '../../auth/user-auth/services/auth-http-header.service';
import { AuthStorageService } from '../../auth/user-auth/services/auth-storage.service';
import { OAuthLibWrapperService } from '../../auth/user-auth/services/oauth-lib-wrapper.service';
import { GlobalMessageService } from '../../global-message/facade/global-message.service';
import { GlobalMessageType } from '../../global-message/models/global-message.model';
import { OccEndpointsService } from '../../occ/services/occ-endpoints.service';
import { InterceptorUtil, USE_CUSTOMER_SUPPORT_AGENT_TOKEN, } from '../../occ/utils/interceptor-util';
import { RoutingService } from '../../routing/facade/routing.service';
import { CsAgentAuthService } from '../facade/csagent-auth.service';
import * as i0 from "@angular/core";
import * as i1 from "../../auth/user-auth/facade/auth.service";
import * as i2 from "../../auth/user-auth/services/auth-storage.service";
import * as i3 from "../facade/csagent-auth.service";
import * as i4 from "../../auth/user-auth/services/oauth-lib-wrapper.service";
import * as i5 from "../../routing/facade/routing.service";
import * as i6 from "../../global-message/facade/global-message.service";
import * as i7 from "../../occ/services/occ-endpoints.service";
/**
 * @deprecated since 3.2, use asm lib instead
 *
 * Overrides `AuthHttpHeaderService` to handle asm calls as well (not only OCC)
 * in cases of normal user session and on customer emulation.
 */
export class AsmAuthHttpHeaderService extends AuthHttpHeaderService {
    constructor(authService, authStorageService, csAgentAuthService, oAuthLibWrapperService, routingService, globalMessageService, occEndpointsService) {
        super(authService, authStorageService, oAuthLibWrapperService, routingService, occEndpointsService, globalMessageService);
        this.authService = authService;
        this.authStorageService = authStorageService;
        this.csAgentAuthService = csAgentAuthService;
        this.oAuthLibWrapperService = oAuthLibWrapperService;
        this.routingService = routingService;
        this.globalMessageService = globalMessageService;
        this.occEndpointsService = occEndpointsService;
    }
    /**
     * @override
     *
     * Checks if particular request should be handled by this service.
     */
    shouldCatchError(request) {
        return (super.shouldCatchError(request) || this.isCSAgentTokenRequest(request));
    }
    /**
     * @override
     *
     * Adds `Authorization` header to occ and CS agent requests.
     * For CS agent requests also removes the `cx-use-csagent-token` header (to avoid problems with CORS).
     */
    alterRequest(request) {
        const hasAuthorizationHeader = !!this.getAuthorizationHeader(request);
        const isCSAgentRequest = this.isCSAgentTokenRequest(request);
        let req = super.alterRequest(request);
        if (!hasAuthorizationHeader && isCSAgentRequest) {
            req = request.clone({
                setHeaders: Object.assign({}, this.createAuthorizationHeader()),
            });
            return InterceptorUtil.removeHeader(USE_CUSTOMER_SUPPORT_AGENT_TOKEN, req);
        }
        return req;
    }
    isCSAgentTokenRequest(request) {
        const isRequestWithCSAgentToken = InterceptorUtil.getInterceptorParam(USE_CUSTOMER_SUPPORT_AGENT_TOKEN, request.headers);
        return Boolean(isRequestWithCSAgentToken);
    }
    /**
     * @override
     *
     * On backend errors indicating expired `refresh_token` we need to logout
     * currently logged in user and CS agent.
     */
    handleExpiredRefreshToken() {
        this.csAgentAuthService
            .isCustomerSupportAgentLoggedIn()
            .pipe(take(1))
            .subscribe((csAgentLoggedIn) => {
            if (csAgentLoggedIn) {
                this.csAgentAuthService.logoutCustomerSupportAgent();
                this.globalMessageService.add({
                    key: 'asm.csagentTokenExpired',
                }, GlobalMessageType.MSG_TYPE_ERROR);
            }
            else {
                super.handleExpiredRefreshToken();
            }
        });
    }
}
AsmAuthHttpHeaderService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AsmAuthHttpHeaderService_Factory() { return new AsmAuthHttpHeaderService(i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i2.AuthStorageService), i0.ɵɵinject(i3.CsAgentAuthService), i0.ɵɵinject(i4.OAuthLibWrapperService), i0.ɵɵinject(i5.RoutingService), i0.ɵɵinject(i6.GlobalMessageService), i0.ɵɵinject(i7.OccEndpointsService)); }, token: AsmAuthHttpHeaderService, providedIn: "root" });
AsmAuthHttpHeaderService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AsmAuthHttpHeaderService.ctorParameters = () => [
    { type: AuthService },
    { type: AuthStorageService },
    { type: CsAgentAuthService },
    { type: OAuthLibWrapperService },
    { type: RoutingService },
    { type: GlobalMessageService },
    { type: OccEndpointsService }
];
