// @spartacus/storefront/esm2015/cms-components/cart/cart-totals/cart-totals.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { CartCouponModule } from '../cart-coupon/cart-coupon.module';
import { CartSharedModule } from '../cart-shared/cart-shared.module';
import { CartTotalsComponent } from './cart-totals.component';
export class CartTotalsModule {
}
CartTotalsModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    UrlModule,
                    CartSharedModule,
                    I18nModule,
                    CartCouponModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            CartTotalsComponent: {
                                component: CartTotalsComponent,
                            },
                        },
                    }),
                ],
                declarations: [CartTotalsComponent],
                exports: [CartTotalsComponent],
                entryComponents: [CartTotalsComponent],
            },] }
];
