// @spartacus/storefront/esm2015/cms-components/product/product-variants/guards/product-variant.guard.js
import { Injectable } from '@angular/core';
import { Router, } from '@angular/router';
import { ProductService, SemanticPathService, } from '@spartacus/core';
import { of } from 'rxjs';
import { filter, map, switchMap, take } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "@angular/router";
/**
 * @deprecated since 3.2
 * Use feature-library @spartacus/product/variants/components instead
 */
export class ProductVariantGuard {
    constructor(productService, semanticPathService, router) {
        this.productService = productService;
        this.semanticPathService = semanticPathService;
        this.router = router;
    }
    canActivate(activatedRoute) {
        var _a;
        const productCode = (_a = activatedRoute.params) === null || _a === void 0 ? void 0 : _a.productCode;
        if (!productCode) {
            return of(true);
        }
        return this.productService.get(productCode, "variants" /* VARIANTS */).pipe(filter(Boolean), switchMap((product) => {
            if (!product.purchasable) {
                const variant = this.findVariant(product.variantOptions);
                // below call might looks redundant but in fact this data is going to be loaded anyways
                // we're just calling it earlier and storing
                return this.productService.get(variant.code, "list" /* LIST */).pipe(filter(Boolean), take(1), map((_product) => {
                    return this.router.createUrlTree(this.semanticPathService.transform({
                        cxRoute: 'product',
                        params: _product,
                    }));
                }));
            }
            else {
                return of(true);
            }
        }));
    }
    findVariant(variants) {
        const results = variants.filter((variant) => {
            return variant.stock && variant.stock.stockLevel ? variant : false;
        });
        return !results.length && variants.length ? variants[0] : results[0];
    }
}
ProductVariantGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductVariantGuard_Factory() { return new ProductVariantGuard(i0.ɵɵinject(i1.ProductService), i0.ɵɵinject(i1.SemanticPathService), i0.ɵɵinject(i2.Router)); }, token: ProductVariantGuard, providedIn: "root" });
ProductVariantGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ProductVariantGuard.ctorParameters = () => [
    { type: ProductService },
    { type: SemanticPathService },
    { type: Router }
];
