// @spartacus/core/esm2015/src/user/connectors/customer-coupon/customer-coupon.connector.js
import { Injectable } from '@angular/core';
import { CustomerCouponAdapter } from './customer-coupon.adapter';
import * as i0 from "@angular/core";
import * as i1 from "./customer-coupon.adapter";
export class CustomerCouponConnector {
    constructor(adapter) {
        this.adapter = adapter;
    }
    getCustomerCoupons(userId, pageSize, currentPage, sort) {
        return this.adapter.getCustomerCoupons(userId, pageSize, currentPage, sort);
    }
    turnOnNotification(userId, couponCode) {
        return this.adapter.turnOnNotification(userId, couponCode);
    }
    turnOffNotification(userId, couponCode) {
        return this.adapter.turnOffNotification(userId, couponCode);
    }
    claimCustomerCoupon(userId, couponCode) {
        return this.adapter.claimCustomerCoupon(userId, couponCode);
    }
}
CustomerCouponConnector.ɵprov = i0.ɵɵdefineInjectable({ factory: function CustomerCouponConnector_Factory() { return new CustomerCouponConnector(i0.ɵɵinject(i1.CustomerCouponAdapter)); }, token: CustomerCouponConnector, providedIn: "root" });
CustomerCouponConnector.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
CustomerCouponConnector.ctorParameters = () => [
    { type: CustomerCouponAdapter }
];
