// @spartacus/storefront/esm2015/cms-structure/page/component/page-component.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ComponentWrapperDirective } from './component-wrapper.directive';
import { ComponentHandler } from './handlers/component-handler';
import { DefaultComponentHandler } from './handlers/default-component.handler';
import { LazyComponentHandler } from './handlers/lazy-component.handler';
import { InnerComponentsHostDirective } from './inner-components-host.directive';
export class PageComponentModule {
}
PageComponentModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule],
                providers: [
                    {
                        provide: ComponentHandler,
                        useExisting: DefaultComponentHandler,
                        multi: true,
                    },
                    {
                        provide: ComponentHandler,
                        useExisting: LazyComponentHandler,
                        multi: true,
                    },
                ],
                declarations: [ComponentWrapperDirective, InnerComponentsHostDirective],
                exports: [ComponentWrapperDirective, InnerComponentsHostDirective],
            },] }
];
