// @spartacus/setup/esm2015/recipes/b2b/b2b-storefront.module.js
import { HttpClientModule } from '@angular/common/http';
import { NgModule } from '@angular/core';
import { CostCenterModule, provideConfig, provideDefaultConfig, } from '@spartacus/core';
import { CmsLibModule, defaultCmsContentProviders, layoutConfig, mediaConfig, StorefrontModule, } from '@spartacus/storefront';
import { defaultB2bCheckoutConfig } from './config/default-b2b-checkout-config';
import { defaultB2bOccConfig } from './config/default-b2b-occ-config';
export class B2bStorefrontModule {
    static withConfig(config) {
        return {
            ngModule: B2bStorefrontModule,
            providers: [provideConfig(config)],
        };
    }
}
B2bStorefrontModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    HttpClientModule,
                    StorefrontModule,
                    // the cms lib module contains all components that added in the bundle
                    CmsLibModule,
                    CostCenterModule.forRoot(),
                ],
                providers: [
                    provideDefaultConfig(layoutConfig),
                    provideDefaultConfig(mediaConfig),
                    provideDefaultConfig(defaultB2bOccConfig),
                    provideDefaultConfig(defaultB2bCheckoutConfig),
                    ...defaultCmsContentProviders,
                ],
                exports: [StorefrontModule],
            },] }
];
