// @spartacus/storefront/esm2015/events/home/home-page-event.builder.js
import { Injectable } from '@angular/core';
import { createFrom, EventService, FeatureConfigService, } from '@spartacus/core';
import { filter, map } from 'rxjs/operators';
import { NavigationEvent } from '../navigation/navigation.event';
import { HomePageEvent } from './home-page.events';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
export class HomePageEventBuilder {
    constructor(eventService, 
    // TODO: #10896 - remove this
    /** @deprecated @since 3.1 - this will be removed in 4.0 */ featureConfigService) {
        this.eventService = eventService;
        this.featureConfigService = featureConfigService;
        this.register();
    }
    register() {
        this.eventService.register(HomePageEvent, this.buildHomePageEvent());
    }
    buildHomePageEvent() {
        return this.eventService.get(NavigationEvent).pipe(filter((navigationEvent) => navigationEvent.semanticRoute === 'home'), map((navigationEvent) => createFrom(HomePageEvent, Object.assign(Object.assign({}, this.createDeprecatedPageEvent(navigationEvent)), { navigation: Object.assign({}, navigationEvent) }))));
    }
    createDeprecatedPageEvent(navigationEvent) {
        if (!this.featureConfigService ||
            this.featureConfigService.isLevel('!3.1')) {
            return Object.assign({}, navigationEvent);
        }
        return undefined;
    }
}
HomePageEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function HomePageEventBuilder_Factory() { return new HomePageEventBuilder(i0.ɵɵinject(i1.EventService), i0.ɵɵinject(i1.FeatureConfigService)); }, token: HomePageEventBuilder, providedIn: "root" });
HomePageEventBuilder.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
HomePageEventBuilder.ctorParameters = () => [
    { type: EventService },
    { type: FeatureConfigService }
];
