// @spartacus/storefront/esm2015/cms-components/product/product-list/product-grid-item/product-grid-item.component.js
import { ChangeDetectionStrategy, Component, Input, Optional, } from '@angular/core';
import { ProductListOutlets } from '../../product-outlets.model';
import { ProductListItemContextSource } from '../model/product-list-item-context-source.model';
import { ProductListItemContext } from '../model/product-list-item-context.model';
export class ProductGridItemComponent {
    constructor(productListItemContextSource) {
        this.productListItemContextSource = productListItemContextSource;
        this.ProductListOutlets = ProductListOutlets;
    }
    ngOnChanges(changes) {
        var _a;
        if (changes === null || changes === void 0 ? void 0 : changes.product) {
            (_a = this.productListItemContextSource) === null || _a === void 0 ? void 0 : _a.product$.next(this.product);
        }
    }
}
ProductGridItemComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-product-grid-item',
                template: "<a\n  [routerLink]=\"{ cxRoute: 'product', params: product } | cxUrl\"\n  class=\"cx-product-image-container\"\n  tabindex=\"-1\"\n>\n  <cx-media\n    class=\"cx-product-image\"\n    [container]=\"product.images?.PRIMARY\"\n    format=\"product\"\n    [alt]=\"product.summary\"\n  ></cx-media>\n</a>\n<a\n  [routerLink]=\"{ cxRoute: 'product', params: product } | cxUrl\"\n  class=\"cx-product-name\"\n  [innerHTML]=\"product.nameHtml\"\n></a>\n\n<div class=\"cx-product-rating\">\n  <cx-star-rating\n    *ngIf=\"product.averageRating\"\n    [rating]=\"product?.averageRating\"\n  ></cx-star-rating>\n  <div *ngIf=\"!product.averageRating\">\n    {{ 'productDetails.noReviews' | cxTranslate }}\n  </div>\n</div>\n<div class=\"cx-product-price-container\">\n  <div class=\"cx-product-price\" aria-label=\"Product price\">\n    {{ product.price?.formattedValue }}\n  </div>\n</div>\n\n<div class=\"cx-variant-style-icons\" *ngIf=\"product.variantOptions\">\n  <cx-variant-style-icons\n    [variants]=\"product.variantOptions\"\n  ></cx-variant-style-icons>\n</div>\n\n<ng-template [cxOutlet]=\"ProductListOutlets.ITEM_ACTIONS\">\n  <cx-add-to-cart\n    *ngIf=\"product.stock?.stockLevelStatus !== 'outOfStock'\"\n    [showQuantity]=\"false\"\n    [product]=\"product\"\n  ></cx-add-to-cart>\n</ng-template>\n",
                changeDetection: ChangeDetectionStrategy.OnPush,
                providers: [
                    ProductListItemContextSource,
                    {
                        provide: ProductListItemContext,
                        useExisting: ProductListItemContextSource,
                    },
                ]
            },] }
];
ProductGridItemComponent.ctorParameters = () => [
    { type: ProductListItemContextSource, decorators: [{ type: Optional }] }
];
ProductGridItemComponent.propDecorators = {
    product: [{ type: Input }]
};
