// @spartacus/user/esm2015/profile/components/register/register.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { I18nModule, NotAuthGuard, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { FormErrorsModule, SpinnerModule } from '@spartacus/storefront';
import { RegisterComponent } from './register.component';
export class RegisterComponentModule {
}
RegisterComponentModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    ReactiveFormsModule,
                    RouterModule,
                    UrlModule,
                    I18nModule,
                    SpinnerModule,
                    FormErrorsModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            RegisterCustomerComponent: {
                                component: RegisterComponent,
                                guards: [NotAuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [RegisterComponent],
            },] }
];
