// @spartacus/storefront/esm2015/cms-components/myaccount/order/return-request-list/order-return-request-list.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { AuthGuard, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { ListNavigationModule } from '../../../../shared/components/list-navigation/list-navigation.module';
import { OrderReturnRequestListComponent } from './order-return-request-list.component';
export class ReturnRequestListModule {
}
ReturnRequestListModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    RouterModule,
                    ListNavigationModule,
                    UrlModule,
                    I18nModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            OrderReturnRequestListComponent: {
                                component: OrderReturnRequestListComponent,
                                guards: [AuthGuard],
                            },
                        },
                    }),
                ],
                declarations: [OrderReturnRequestListComponent],
                exports: [OrderReturnRequestListComponent],
                entryComponents: [OrderReturnRequestListComponent],
            },] }
];
