// @spartacus/storefront/esm2015/shared/components/table/table-data-cell/table-data-cell.component.js
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { OutletContextData } from '../../../../cms-structure/outlet/outlet.model';
export class TableDataCellComponent {
    constructor(outlet) {
        this.outlet = outlet;
    }
    get value() {
        return this.model[this.field];
    }
    get model() {
        var _a;
        return (_a = this.outlet) === null || _a === void 0 ? void 0 : _a.context;
    }
    get field() {
        var _a, _b;
        return (_b = (_a = this.outlet) === null || _a === void 0 ? void 0 : _a.context) === null || _b === void 0 ? void 0 : _b._field;
    }
}
TableDataCellComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-table-data-cell',
                template: `{{ value }}`,
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
TableDataCellComponent.ctorParameters = () => [
    { type: OutletContextData }
];
TableDataCellComponent.propDecorators = {
    value: [{ type: HostBinding, args: ['attr.title',] }]
};
