// @spartacus/user/bundles/spartacus-user-profile-root.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/core'), require('@spartacus/core')) :
    typeof define === 'function' && define.amd ? define('@spartacus/user/profile/root', ['exports', '@angular/core', '@spartacus/core'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.spartacus = global.spartacus || {}, global.spartacus.user = global.spartacus.user || {}, global.spartacus.user.profile = global.spartacus.user.profile || {}, global.spartacus.user.profile.root = {}), global.ng.core, global.core));
}(this, (function (exports, i0, core) { 'use strict';

    var USER_PROFILE_FEATURE = 'userProfile';
    var USER_PROFILE_CORE_FEATURE = 'userProfileCore';

    function UserProfileFacadeFactory() {
        return core.facadeFactory({
            facade: UserProfileFacade,
            feature: USER_PROFILE_CORE_FEATURE,
            methods: ['get', 'update', 'close', 'getTitles'],
        });
    }
    var UserProfileFacade = /** @class */ (function () {
        function UserProfileFacade() {
        }
        return UserProfileFacade;
    }());
    UserProfileFacade.ɵprov = i0.ɵɵdefineInjectable({ factory: UserProfileFacadeFactory, token: UserProfileFacade, providedIn: "root" });
    UserProfileFacade.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                    useFactory: UserProfileFacadeFactory,
                },] }
    ];

    function UserRegisterFacadeFactory() {
        return core.facadeFactory({
            facade: UserRegisterFacade,
            feature: USER_PROFILE_CORE_FEATURE,
            methods: ['register', 'registerGuest', 'getTitles'],
        });
    }
    var UserRegisterFacade = /** @class */ (function () {
        function UserRegisterFacade() {
        }
        return UserRegisterFacade;
    }());
    UserRegisterFacade.ɵprov = i0.ɵɵdefineInjectable({ factory: UserRegisterFacadeFactory, token: UserRegisterFacade, providedIn: "root" });
    UserRegisterFacade.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                    useFactory: UserRegisterFacadeFactory,
                },] }
    ];

    // TODO: Inline this factory when we start releasing Ivy compiled libraries
    function defaultUserProfileComponentsConfig() {
        var _a;
        var config = {
            featureModules: (_a = {},
                _a[USER_PROFILE_FEATURE] = {
                    cmsComponents: [
                        'RegisterCustomerComponent',
                        'UpdateProfileComponent',
                        'UpdateEmailComponent',
                        'UpdatePasswordComponent',
                        'ForgotPasswordComponent',
                        'ResetPasswordComponent',
                        'CloseAccountComponent',
                    ],
                },
                // by default core is bundled together with components
                _a[USER_PROFILE_CORE_FEATURE] = USER_PROFILE_FEATURE,
                _a),
        };
        return config;
    }
    var UserProfileRootModule = /** @class */ (function () {
        function UserProfileRootModule() {
        }
        return UserProfileRootModule;
    }());
    UserProfileRootModule.decorators = [
        { type: i0.NgModule, args: [{
                    providers: [
                        core.provideDefaultConfigFactory(defaultUserProfileComponentsConfig),
                        {
                            provide: core.UserProfileFacadeTransitionalToken,
                            useExisting: UserProfileFacade,
                        },
                        {
                            provide: core.UserRegisterFacadeTransitionalToken,
                            useExisting: UserRegisterFacade,
                        },
                    ],
                },] }
    ];

    function UserEmailFacadeFactory() {
        return core.facadeFactory({
            facade: UserEmailFacade,
            feature: USER_PROFILE_CORE_FEATURE,
            methods: ['update'],
        });
    }
    var UserEmailFacade = /** @class */ (function () {
        function UserEmailFacade() {
        }
        return UserEmailFacade;
    }());
    UserEmailFacade.ɵprov = i0.ɵɵdefineInjectable({ factory: UserEmailFacadeFactory, token: UserEmailFacade, providedIn: "root" });
    UserEmailFacade.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                    useFactory: UserEmailFacadeFactory,
                },] }
    ];

    function UserPasswordFacadeFactory() {
        return core.facadeFactory({
            facade: UserPasswordFacade,
            feature: USER_PROFILE_CORE_FEATURE,
            methods: ['update', 'reset', 'requestForgotPasswordEmail'],
        });
    }
    var UserPasswordFacade = /** @class */ (function () {
        function UserPasswordFacade() {
        }
        return UserPasswordFacade;
    }());
    UserPasswordFacade.ɵprov = i0.ɵɵdefineInjectable({ factory: UserPasswordFacadeFactory, token: UserPasswordFacade, providedIn: "root" });
    UserPasswordFacade.decorators = [
        { type: i0.Injectable, args: [{
                    providedIn: 'root',
                    useFactory: UserPasswordFacadeFactory,
                },] }
    ];

    /** AUGMENTABLE_TYPES_END */

    /**
     * Generated bundle index. Do not edit.
     */

    exports.USER_PROFILE_CORE_FEATURE = USER_PROFILE_CORE_FEATURE;
    exports.USER_PROFILE_FEATURE = USER_PROFILE_FEATURE;
    exports.UserEmailFacade = UserEmailFacade;
    exports.UserEmailFacadeFactory = UserEmailFacadeFactory;
    exports.UserPasswordFacade = UserPasswordFacade;
    exports.UserPasswordFacadeFactory = UserPasswordFacadeFactory;
    exports.UserProfileFacade = UserProfileFacade;
    exports.UserProfileFacadeFactory = UserProfileFacadeFactory;
    exports.UserProfileRootModule = UserProfileRootModule;
    exports.UserRegisterFacade = UserRegisterFacade;
    exports.UserRegisterFacadeFactory = UserRegisterFacadeFactory;
    exports.defaultUserProfileComponentsConfig = defaultUserProfileComponentsConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
