// @spartacus/user/bundles/spartacus-user-profile-occ.umd.js
(function (global, factory) {
    typeof exports === 'object' && typeof module !== 'undefined' ? factory(exports, require('@angular/common/http'), require('@angular/core'), require('@spartacus/core'), require('@spartacus/user/profile/core'), require('rxjs'), require('rxjs/operators')) :
    typeof define === 'function' && define.amd ? define('@spartacus/user/profile/occ', ['exports', '@angular/common/http', '@angular/core', '@spartacus/core', '@spartacus/user/profile/core', 'rxjs', 'rxjs/operators'], factory) :
    (global = typeof globalThis !== 'undefined' ? globalThis : global || self, factory((global.spartacus = global.spartacus || {}, global.spartacus.user = global.spartacus.user || {}, global.spartacus.user.profile = global.spartacus.user.profile || {}, global.spartacus.user.profile.occ = {}), global.ng.common.http, global.ng.core, global.core, global.spartacus.user.profile.core, global.rxjs, global.rxjs.operators));
}(this, (function (exports, http, core$2, core$1, core, rxjs, operators) { 'use strict';

    var defaultOccUserProfileConfig = {
        backend: {
            occ: {
                endpoints: {
                    userRegister: 'users',
                    userForgotPassword: 'forgottenpasswordtokens',
                    userResetPassword: 'resetpassword',
                    userUpdateLoginId: 'users/${userId}/login',
                    userUpdatePassword: 'users/${userId}/password',
                    titles: 'titles',
                },
            },
        },
    };

    var OccUserProfileAdapter = /** @class */ (function () {
        function OccUserProfileAdapter(http, occEndpoints, converter) {
            this.http = http;
            this.occEndpoints = occEndpoints;
            this.converter = converter;
        }
        OccUserProfileAdapter.prototype.update = function (userId, user) {
            var endpoint = this.occEndpoints.isConfigured('userUpdateProfile')
                ? 'userUpdateProfile'
                : 'user';
            var url = this.occEndpoints.getUrl(endpoint, { userId: userId });
            user = this.converter.convert(user, core.USER_PROFILE_SERIALIZER);
            return this.http
                .patch(url, user)
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.register = function (user) {
            var url = this.occEndpoints.getUrl('userRegister');
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/json',
            });
            headers = core$1.InterceptorUtil.createHeader(core$1.USE_CLIENT_TOKEN, true, headers);
            user = this.converter.convert(user, core.USER_SIGN_UP_SERIALIZER);
            return this.http
                .post(url, user, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }), this.converter.pipeable(core.USER_PROFILE_NORMALIZER));
        };
        OccUserProfileAdapter.prototype.registerGuest = function (guid, password) {
            var url = this.occEndpoints.getUrl('userRegister');
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/x-www-form-urlencoded',
            });
            headers = core$1.InterceptorUtil.createHeader(core$1.USE_CLIENT_TOKEN, true, headers);
            var httpParams = new http.HttpParams()
                .set('guid', guid)
                .set('password', password);
            return this.http
                .post(url, httpParams, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }), this.converter.pipeable(core.USER_PROFILE_NORMALIZER));
        };
        OccUserProfileAdapter.prototype.requestForgotPasswordEmail = function (userEmailAddress) {
            var url = this.occEndpoints.getUrl('userForgotPassword');
            var httpParams = new http.HttpParams().set('userId', userEmailAddress);
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/x-www-form-urlencoded',
            });
            headers = core$1.InterceptorUtil.createHeader(core$1.USE_CLIENT_TOKEN, true, headers);
            return this.http
                .post(url, httpParams, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.resetPassword = function (token, newPassword) {
            var url = this.occEndpoints.getUrl('userResetPassword');
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/json',
            });
            headers = core$1.InterceptorUtil.createHeader(core$1.USE_CLIENT_TOKEN, true, headers);
            return this.http
                .post(url, { token: token, newPassword: newPassword }, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.updateEmail = function (userId, currentPassword, newUserId) {
            var url = this.occEndpoints.getUrl('userUpdateLoginId', { userId: userId });
            var httpParams = new http.HttpParams()
                .set('password', currentPassword)
                .set('newLogin', newUserId);
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/x-www-form-urlencoded',
            });
            return this.http
                .put(url, httpParams, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.updatePassword = function (userId, oldPassword, newPassword) {
            var url = this.occEndpoints.getUrl('userUpdatePassword', { userId: userId });
            var httpParams = new http.HttpParams()
                .set('old', oldPassword)
                .set('new', newPassword);
            var headers = new http.HttpHeaders({
                'Content-Type': 'application/x-www-form-urlencoded',
            });
            return this.http
                .put(url, httpParams, { headers: headers })
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.close = function (userId) {
            var endpoint = this.occEndpoints.isConfigured('userCloseAccount')
                ? 'userCloseAccount'
                : 'user';
            var url = this.occEndpoints.getUrl(endpoint, { userId: userId });
            return this.http
                .delete(url)
                .pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }));
        };
        OccUserProfileAdapter.prototype.loadTitles = function () {
            var url = this.occEndpoints.getUrl('titles');
            return this.http.get(url).pipe(operators.catchError(function (error) { return rxjs.throwError(core$1.normalizeHttpError(error)); }), operators.map(function (titleList) { var _a; return (_a = titleList.titles) !== null && _a !== void 0 ? _a : []; }), this.converter.pipeableMany(core.TITLE_NORMALIZER));
        };
        return OccUserProfileAdapter;
    }());
    OccUserProfileAdapter.decorators = [
        { type: core$2.Injectable }
    ];
    OccUserProfileAdapter.ctorParameters = function () { return [
        { type: http.HttpClient },
        { type: core$1.OccEndpointsService },
        { type: core$1.ConverterService }
    ]; };

    var UserProfileOccModule = /** @class */ (function () {
        function UserProfileOccModule() {
        }
        return UserProfileOccModule;
    }());
    UserProfileOccModule.decorators = [
        { type: core$2.NgModule, args: [{
                    providers: [
                        core$1.provideDefaultConfig(defaultOccUserProfileConfig),
                        { provide: core.UserProfileAdapter, useClass: OccUserProfileAdapter },
                    ],
                },] }
    ];

    /**
     * Generated bundle index. Do not edit.
     */

    exports.OccUserProfileAdapter = OccUserProfileAdapter;
    exports.UserProfileOccModule = UserProfileOccModule;
    exports.defaultOccUserProfileConfig = defaultOccUserProfileConfig;

    Object.defineProperty(exports, '__esModule', { value: true });

})));
