// @spartacus/core/esm2015/src/occ/adapters/checkout/occ-checkout-payment-type.adapter.js
import { HttpClient, HttpParams } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { map } from 'rxjs/operators';
import { CART_NORMALIZER } from '../../../cart/connectors/cart/converters';
import { PAYMENT_TYPE_NORMALIZER } from '../../../checkout/connectors/payment-type/converters';
import { ConverterService } from '../../../util/converter.service';
import { OccEndpointsService } from '../../services/occ-endpoints.service';
const ENDPOINT_PAYMENT_TYPES = 'paymenttypes';
export class OccCheckoutPaymentTypeAdapter {
    constructor(http, occEndpoints, converter) {
        this.http = http;
        this.occEndpoints = occEndpoints;
        this.converter = converter;
    }
    loadPaymentTypes() {
        return this.http
            .get(this.occEndpoints.getEndpoint(ENDPOINT_PAYMENT_TYPES))
            .pipe(map((paymentTypeList) => paymentTypeList.paymentTypes), this.converter.pipeableMany(PAYMENT_TYPE_NORMALIZER));
    }
    setPaymentType(userId, cartId, paymentType, purchaseOrderNumber) {
        let httpParams = new HttpParams().set('paymentType', paymentType);
        if (purchaseOrderNumber !== undefined) {
            httpParams = httpParams.set('purchaseOrderNumber', purchaseOrderNumber);
        }
        /* eslint-disable max-len */
        httpParams = httpParams.set('fields', 'DEFAULT,potentialProductPromotions,appliedProductPromotions,potentialOrderPromotions,appliedOrderPromotions,entries(totalPrice(formattedValue),product(images(FULL),stock(FULL)),basePrice(formattedValue,value),updateable),totalPrice(formattedValue),totalItems,totalPriceWithTax(formattedValue),totalDiscounts(value,formattedValue),subTotal(formattedValue),deliveryItemsQuantity,deliveryCost(formattedValue),totalTax(formattedValue, value),pickupItemsQuantity,net,appliedVouchers,productDiscounts(formattedValue),user');
        // TODO(#8877): Should we improve configurable endpoints for this use case?
        return this.http
            .put(this.getCartEndpoint(userId) + cartId + '/paymenttype', {}, {
            params: httpParams,
        })
            .pipe(this.converter.pipeable(CART_NORMALIZER));
    }
    getCartEndpoint(userId) {
        const cartEndpoint = 'users/' + userId + '/carts/';
        return this.occEndpoints.getEndpoint(cartEndpoint);
    }
}
OccCheckoutPaymentTypeAdapter.decorators = [
    { type: Injectable }
];
OccCheckoutPaymentTypeAdapter.ctorParameters = () => [
    { type: HttpClient },
    { type: OccEndpointsService },
    { type: ConverterService }
];
