// @spartacus/storefront/esm2015/cms-components/cart/add-to-wishlist/add-to-wish-list.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { RouterModule } from '@angular/router';
import { I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { IconModule } from '../../misc/icon/icon.module';
import { AddToWishListComponent } from './add-to-wish-list.component';
export class AddToWishListModule {
}
AddToWishListModule.decorators = [
    { type: NgModule, args: [{
                imports: [CommonModule, I18nModule, IconModule, RouterModule, UrlModule],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            AddToWishListComponent: {
                                component: AddToWishListComponent,
                            },
                        },
                    }),
                ],
                declarations: [AddToWishListComponent],
                entryComponents: [AddToWishListComponent],
                exports: [AddToWishListComponent],
            },] }
];
