// @spartacus/setup/esm2015/recipes/b2b/config/default-b2b-occ-config.js
export const defaultB2bOccConfig = {
    backend: {
        occ: {
            endpoints: {
                user: 'orgUsers/${userId}',
                userUpdateProfile: 'users/${userId}',
                userCloseAccount: 'users/${userId}',
                addEntries: 'orgUsers/${userId}/carts/${cartId}/entries?quantity=${quantity}',
                setDeliveryAddress: 'orgUsers/${userId}/carts/${cartId}/addresses/delivery',
                placeOrder: 'orgUsers/${userId}/orders?fields=FULL',
                scheduleReplenishmentOrder: 'orgUsers/${userId}/replenishmentOrders?fields=FULL,costCenter(FULL),purchaseOrderNumber,paymentType',
                replenishmentOrderDetails: 'users/${userId}/replenishmentOrders/${replenishmentOrderCode}?fields=FULL,costCenter(FULL),purchaseOrderNumber,paymentType,user',
                replenishmentOrderDetailsHistory: 'users/${userId}/replenishmentOrders/${replenishmentOrderCode}/orders',
                cancelReplenishmentOrder: 'users/${userId}/replenishmentOrders/${replenishmentOrderCode}?fields=FULL,costCenter(FULL),purchaseOrderNumber,paymentType,user',
                replenishmentOrderHistory: 'users/${userId}/replenishmentOrders?fields=FULL,replenishmentOrders(FULL, purchaseOrderNumber)',
            },
        },
    },
};
