// @spartacus/storefront/esm2015/events/page/page-event.builder.js
import { Injectable } from '@angular/core';
import { ActionsSubject } from '@ngrx/store';
import { createFrom, EventService, FeatureConfigService, } from '@spartacus/core';
import { map } from 'rxjs/operators';
import { NavigationEvent } from '../navigation/navigation.event';
import { PageEvent } from './page.events';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
import * as i2 from "@spartacus/core";
// TODO: #10896 - delete this whole file
/**
 * @deprecated @since 3.1 - this builder will be removed in 4.0. Please use NavigationEventBuilder and NavigationEvent instead.
 */
export class PageEventBuilder {
    constructor(
    /** @deprecated @since 3.1 - this will be removed in 4.0 */
    actions, eventService, 
    /** @deprecated @since 3.1 - this will be removed in 4.0 */ featureConfigService) {
        this.actions = actions;
        this.eventService = eventService;
        this.featureConfigService = featureConfigService;
        this.register();
    }
    register() {
        var _a;
        if ((_a = this.featureConfigService) === null || _a === void 0 ? void 0 : _a.isLevel('!3.1')) {
            this.eventService.register(PageEvent, this.buildPageEvent());
        }
    }
    /**
     * @deprecated @since 3.1 - this will be removed in 4.0. Please use `NavigationEvent` instead.
     */
    buildPageEvent() {
        return this.eventService.get(NavigationEvent).pipe(map((navigationEvent) => createFrom(PageEvent, {
            context: navigationEvent.context,
            semanticRoute: navigationEvent.semanticRoute,
            url: navigationEvent.url,
            params: navigationEvent.params,
            navigation: Object.assign({}, navigationEvent),
        })));
    }
}
PageEventBuilder.ɵprov = i0.ɵɵdefineInjectable({ factory: function PageEventBuilder_Factory() { return new PageEventBuilder(i0.ɵɵinject(i1.ActionsSubject), i0.ɵɵinject(i2.EventService), i0.ɵɵinject(i2.FeatureConfigService)); }, token: PageEventBuilder, providedIn: "root" });
PageEventBuilder.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
PageEventBuilder.ctorParameters = () => [
    { type: ActionsSubject },
    { type: EventService },
    { type: FeatureConfigService }
];
