// @spartacus/storefront/esm2015/cms-components/user/logout/logout.guard.js
import { Injectable } from '@angular/core';
import { Router } from '@angular/router';
import { AuthRedirectService, AuthService, CmsService, PageType, ProtectedRoutesService, SemanticPathService, } from '@spartacus/core';
import { from } from 'rxjs';
import { map, switchMap } from 'rxjs/operators';
import * as i0 from "@angular/core";
import * as i1 from "@spartacus/core";
import * as i2 from "@angular/router";
/**
 * Guards the _logout_ route.
 *
 * Takes care of routing the user to a logout page (if available) or redirects to
 * the homepage. If the homepage is protected, the user is redirected
 * to the login route instead.
 */
export class LogoutGuard {
    constructor(auth, cms, semanticPathService, protectedRoutes, router, 
    // TODO(#11380): Make service required
    authRedirectService) {
        this.auth = auth;
        this.cms = cms;
        this.semanticPathService = semanticPathService;
        this.protectedRoutes = protectedRoutes;
        this.router = router;
        this.authRedirectService = authRedirectService;
    }
    canActivate() {
        var _a;
        // TODO(#11380): Remove the optional chaining
        // Logout route should never be remembered as a redirect url after login (that would cause logout right after login).
        (_a = this.authRedirectService) === null || _a === void 0 ? void 0 : _a.reportNotAuthGuard();
        /**
         * First we want to complete logout process before redirecting to logout page
         * We want to avoid errors like `token is no longer valid`
         */
        return from(this.logout()).pipe(switchMap(() => {
            return this.cms
                .hasPage({
                id: this.semanticPathService.get('logout'),
                type: PageType.CONTENT_PAGE,
            })
                .pipe(map((hasPage) => {
                if (!hasPage) {
                    return this.getRedirectUrl();
                }
                // TODO(#9385): Use CMS page guard here.
                return hasPage;
            }));
        }));
    }
    logout() {
        return this.auth.coreLogout();
    }
    /**
     * Whenever there is no specific "logout" page configured in the CMS,
     * we redirect after the user is logged out.
     *
     * The user gets redirected to the homepage, unless the homepage is protected
     * (in case of a closed shop). We'll redirect to the login page instead.
     */
    getRedirectUrl() {
        const cxRoute = this.protectedRoutes.shouldProtect ? 'login' : 'home';
        return this.router.parseUrl(this.semanticPathService.get(cxRoute));
    }
}
LogoutGuard.ɵprov = i0.ɵɵdefineInjectable({ factory: function LogoutGuard_Factory() { return new LogoutGuard(i0.ɵɵinject(i1.AuthService), i0.ɵɵinject(i1.CmsService), i0.ɵɵinject(i1.SemanticPathService), i0.ɵɵinject(i1.ProtectedRoutesService), i0.ɵɵinject(i2.Router), i0.ɵɵinject(i1.AuthRedirectService)); }, token: LogoutGuard, providedIn: "root" });
LogoutGuard.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
LogoutGuard.ctorParameters = () => [
    { type: AuthService },
    { type: CmsService },
    { type: SemanticPathService },
    { type: ProtectedRoutesService },
    { type: Router },
    { type: AuthRedirectService }
];
