// @spartacus/core/esm2015/src/cart/store/reducers/index.js
import { InjectionToken } from '@angular/core';
import { AuthActions } from '../../../auth/user-auth/store/actions/index';
import { entityProcessesLoaderReducer } from '../../../state/utils/entity-processes-loader/entity-processes-loader.reducer';
import { MULTI_CART_DATA } from '../multi-cart-state';
import { activeCartReducer, cartEntitiesReducer, wishListReducer, } from './multi-cart.reducer';
export function clearMultiCartState(reducer) {
    return function (state, action) {
        if (action.type === AuthActions.LOGOUT) {
            state = undefined;
        }
        return reducer(state, action);
    };
}
export const multiCartMetaReducers = [clearMultiCartState];
export const multiCartReducerToken = new InjectionToken('MultiCartReducers');
export function getMultiCartReducers() {
    return {
        carts: entityProcessesLoaderReducer(MULTI_CART_DATA, cartEntitiesReducer),
        active: activeCartReducer,
        wishList: wishListReducer,
    };
}
export const multiCartReducerProvider = {
    provide: multiCartReducerToken,
    useFactory: getMultiCartReducers,
};
