// @spartacus/user/esm2015/profile/components/update-profile/update-profile.module.js
import { CommonModule } from '@angular/common';
import { NgModule } from '@angular/core';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { RouterModule } from '@angular/router';
import { AuthGuard, GlobalMessageService, I18nModule, provideDefaultConfig, UrlModule, } from '@spartacus/core';
import { FormErrorsModule, SpinnerModule } from '@spartacus/storefront';
import { UserProfileFacade } from '@spartacus/user/profile/root';
import { UpdateProfileComponentService } from './update-profile-component.service';
import { UpdateProfileComponent } from './update-profile.component';
export class UpdateProfileModule {
}
UpdateProfileModule.decorators = [
    { type: NgModule, args: [{
                imports: [
                    CommonModule,
                    FormsModule,
                    ReactiveFormsModule,
                    SpinnerModule,
                    I18nModule,
                    FormErrorsModule,
                    RouterModule,
                    UrlModule,
                ],
                providers: [
                    provideDefaultConfig({
                        cmsComponents: {
                            UpdateProfileComponent: {
                                component: UpdateProfileComponent,
                                guards: [AuthGuard],
                                providers: [
                                    {
                                        provide: UpdateProfileComponentService,
                                        useClass: UpdateProfileComponentService,
                                        deps: [UserProfileFacade, GlobalMessageService],
                                    },
                                ],
                            },
                        },
                    }),
                ],
                declarations: [UpdateProfileComponent],
            },] }
];
