// @spartacus/core/esm2015/src/checkout/http-interceptors/checkout-cart.interceptor.js
import { HttpErrorResponse, } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { throwError } from 'rxjs';
import { catchError, switchMap, take } from 'rxjs/operators';
import { MultiCartService } from '../../cart/facade/multi-cart.service';
import { RoutingService } from '../../routing/index';
import * as i0 from "@angular/core";
import * as i1 from "../../routing/facade/routing.service";
import * as i2 from "../../cart/facade/multi-cart.service";
/**
 * Interceptor that handles "Cart not found" errors while a user is in a checkout step.
 *
 * When a user doing a checkout has a "Cart not found" error, he is redirected to checkout and the cart is reloaded.
 * If a "Cart not found" error happens and the user is not on checkout, this interceptor does not perform any actions.
 */
export class CheckoutCartInterceptor {
    constructor(routingService, multiCartService) {
        this.routingService = routingService;
        this.multiCartService = multiCartService;
    }
    intercept(request, next) {
        return this.routingService.getRouterState().pipe(take(1), switchMap((state) => {
            return next.handle(request).pipe(catchError((response) => {
                var _a;
                if (response instanceof HttpErrorResponse &&
                    this.isUserInCheckoutRoute((_a = state.state) === null || _a === void 0 ? void 0 : _a.semanticRoute)) {
                    if (this.isCartNotFoundError(response)) {
                        this.routingService.go({ cxRoute: 'cart' });
                        const cartCode = this.getCartIdFromError(response);
                        if (cartCode) {
                            this.multiCartService.reloadCart(cartCode);
                        }
                    }
                }
                return throwError(response);
            }));
        }));
    }
    /**
     * Returns true if the parameter semantic route is part of "checkout"
     * Checkout semantic routes:
     * checkout
     * checkoutPaymentType
     * CheckoutShippingAddress
     * checkoutDeliveryMode
     * checkoutPaymentDetails
     * checkoutReviewOrder
     * checkoutLogin
     * @param semanticRoute
     */
    isUserInCheckoutRoute(semanticRoute) {
        return semanticRoute === null || semanticRoute === void 0 ? void 0 : semanticRoute.toLowerCase().startsWith('checkout');
    }
    /**
     * Checks of the error is for a cart not found, i.e. the cart doesn't exist anymore
     *
     * @param response
     */
    isCartNotFoundError(response) {
        var _a, _b, _c, _d, _e, _f;
        return (response.status === 400 &&
            ((_c = (_b = (_a = response.error) === null || _a === void 0 ? void 0 : _a.errors) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.type) === 'CartError' &&
            ((_f = (_e = (_d = response.error) === null || _d === void 0 ? void 0 : _d.errors) === null || _e === void 0 ? void 0 : _e[0]) === null || _f === void 0 ? void 0 : _f.reason) === 'notFound');
    }
    getCartIdFromError(response) {
        var _a, _b, _c;
        return (_c = (_b = (_a = response.error) === null || _a === void 0 ? void 0 : _a.errors) === null || _b === void 0 ? void 0 : _b[0]) === null || _c === void 0 ? void 0 : _c.subject;
    }
}
CheckoutCartInterceptor.ɵprov = i0.ɵɵdefineInjectable({ factory: function CheckoutCartInterceptor_Factory() { return new CheckoutCartInterceptor(i0.ɵɵinject(i1.RoutingService), i0.ɵɵinject(i2.MultiCartService)); }, token: CheckoutCartInterceptor, providedIn: "root" });
CheckoutCartInterceptor.decorators = [
    { type: Injectable, args: [{ providedIn: 'root' },] }
];
CheckoutCartInterceptor.ctorParameters = () => [
    { type: RoutingService },
    { type: MultiCartService }
];
