// @spartacus/storefront/esm2015/cms-components/content/banner/banner.component.js
import { ChangeDetectionStrategy, Component, HostBinding } from '@angular/core';
import { tap } from 'rxjs/operators';
import { CmsComponentData } from '../../../cms-structure/page/model/cms-component-data';
export class BannerComponent {
    constructor(component) {
        this.component = component;
        this.data$ = this.component.data$.pipe(tap((data) => (this.styleClasses = data.styleClasses)));
    }
    /**
     * Returns `_blank` to force opening the link in a new window whenever the
     * `data.external` flag is set to true.
     */
    getTarget(data) {
        return data.external === 'true' || data.external === true ? '_blank' : null;
    }
}
BannerComponent.decorators = [
    { type: Component, args: [{
                selector: 'cx-banner',
                template: "<cx-generic-link\n  *ngIf=\"data$ | async as data\"\n  [url]=\"data.urlLink\"\n  [target]=\"getTarget(data)\"\n>\n  <p class=\"headline\" *ngIf=\"data.headline\" [innerHTML]=\"data.headline\"></p>\n  <cx-media [container]=\"data.media\"></cx-media>\n  <p class=\"content\" *ngIf=\"data.content\" [innerHTML]=\"data.content\"></p>\n</cx-generic-link>\n",
                changeDetection: ChangeDetectionStrategy.OnPush
            },] }
];
BannerComponent.ctorParameters = () => [
    { type: CmsComponentData }
];
BannerComponent.propDecorators = {
    styleClasses: [{ type: HostBinding, args: ['class',] }]
};
