// @spartacus/core/esm2015/src/product/facade/product-reference.service.js
import { Injectable } from '@angular/core';
import { select, Store } from '@ngrx/store';
import { ProductSelectors } from '../store';
import { ProductActions } from '../store/actions/index';
import * as i0 from "@angular/core";
import * as i1 from "@ngrx/store";
export class ProductReferenceService {
    constructor(store) {
        this.store = store;
    }
    loadProductReferences(productCode, referenceType, pageSize) {
        this.store.dispatch(new ProductActions.LoadProductReferences({
            productCode,
            referenceType,
            pageSize,
        }));
    }
    getProductReferences(productCode, referenceType) {
        return this.store.pipe(select(ProductSelectors.getSelectedProductReferencesFactory(productCode, referenceType)));
    }
    cleanReferences() {
        this.store.dispatch(new ProductActions.CleanProductReferences());
    }
}
ProductReferenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function ProductReferenceService_Factory() { return new ProductReferenceService(i0.ɵɵinject(i1.Store)); }, token: ProductReferenceService, providedIn: "root" });
ProductReferenceService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
ProductReferenceService.ctorParameters = () => [
    { type: Store }
];
