// @spartacus/core/esm2015/src/auth/user-auth/services/auth-state-persistence.service.js
import { Injectable } from '@angular/core';
import { combineLatest, Subscription } from 'rxjs';
import { filter, map } from 'rxjs/operators';
import { StatePersistenceService } from '../../../state/services/state-persistence.service';
import { UserIdService } from '../facade/user-id.service';
import { AuthRedirectStorageService } from './auth-redirect-storage.service';
import { AuthStorageService } from './auth-storage.service';
import * as i0 from "@angular/core";
import * as i1 from "../../../state/services/state-persistence.service";
import * as i2 from "../facade/user-id.service";
import * as i3 from "./auth-storage.service";
import * as i4 from "./auth-redirect-storage.service";
/**
 * Responsible for saving the authorization data (userId, token, redirectUrl) in browser storage.
 */
export class AuthStatePersistenceService {
    constructor(statePersistenceService, userIdService, authStorageService, authRedirectStorageService) {
        this.statePersistenceService = statePersistenceService;
        this.userIdService = userIdService;
        this.authStorageService = authStorageService;
        this.authRedirectStorageService = authRedirectStorageService;
        this.subscription = new Subscription();
        /**
         * Identifier used for storage key.
         */
        this.key = 'auth';
    }
    /**
     * Initializes the synchronization between state and browser storage.
     */
    initSync() {
        this.subscription.add(this.statePersistenceService.syncWithStorage({
            key: this.key,
            state$: this.getAuthState(),
            onRead: (state) => this.onRead(state),
        }));
    }
    /**
     * Gets and transforms state from different sources into the form that should
     * be saved in storage.
     */
    getAuthState() {
        return combineLatest([
            this.authStorageService.getToken().pipe(filter((state) => !!state), map((state) => {
                return Object.assign({}, state);
            })),
            this.userIdService.getUserId(),
            this.authRedirectStorageService.getRedirectUrl(),
        ]).pipe(map(([authToken, userId, redirectUrl]) => {
            let token = authToken;
            if (token) {
                token = Object.assign({}, token);
                // To minimize risk of user account hijacking we don't persist user refresh_token
                delete token.refresh_token;
            }
            return { token, userId, redirectUrl };
        }));
    }
    /**
     * Function called on each browser storage read.
     * Used to update state from browser -> state.
     */
    onRead(state) {
        if (state === null || state === void 0 ? void 0 : state.token) {
            this.authStorageService.setToken(state.token);
        }
        if (state === null || state === void 0 ? void 0 : state.redirectUrl) {
            this.authRedirectStorageService.setRedirectUrl(state.redirectUrl);
        }
        if (state === null || state === void 0 ? void 0 : state.userId) {
            this.userIdService.setUserId(state.userId);
        }
        else {
            this.userIdService.clearUserId();
        }
    }
    /**
     * Reads synchronously state from storage and returns it.
     */
    readStateFromStorage() {
        return this.statePersistenceService.readStateFromStorage({
            key: this.key,
        });
    }
    /**
     * Check synchronously in browser storage if user is logged in (required by transfer state reducer).
     * For most cases `isUserLoggedIn` from the `AuthService` should be used instead of this.
     */
    isUserLoggedIn() {
        var _a, _b;
        return Boolean((_b = (_a = this.readStateFromStorage()) === null || _a === void 0 ? void 0 : _a.token) === null || _b === void 0 ? void 0 : _b.access_token);
    }
    ngOnDestroy() {
        this.subscription.unsubscribe();
    }
}
AuthStatePersistenceService.ɵprov = i0.ɵɵdefineInjectable({ factory: function AuthStatePersistenceService_Factory() { return new AuthStatePersistenceService(i0.ɵɵinject(i1.StatePersistenceService), i0.ɵɵinject(i2.UserIdService), i0.ɵɵinject(i3.AuthStorageService), i0.ɵɵinject(i4.AuthRedirectStorageService)); }, token: AuthStatePersistenceService, providedIn: "root" });
AuthStatePersistenceService.decorators = [
    { type: Injectable, args: [{
                providedIn: 'root',
            },] }
];
AuthStatePersistenceService.ctorParameters = () => [
    { type: StatePersistenceService },
    { type: UserIdService },
    { type: AuthStorageService },
    { type: AuthRedirectStorageService }
];
